---
title: "Getting Started with Zorn"
output:
  html_document:
    theme: united
    df_print: kable
  pdf_document: default
---

Zorn and Bascet can do many things:

* Preprocess reads for single-cell metagenomics, WGS and RNA-seq
* Perform de novo assembly and other tasks relevant for whole-genome sequencing
* Perform custom per-cell detailed analysis operations

The interface is designed to interoperate with Signac and Seurat for high-level
analysis of the data, after the preprocessing.


Depending on your use case, you will need different tutorials. However,
these are in common for all workflows:

* [Installing Zorn and Bascet](install.html)
* [Using SLURM](slurm.html) - If you want to run this in a high-performance environment
* [Debarcoding tutorial](debarcoding.html) - Ingesting raw FASTQ, trimming, and preparing for later steps
* [Read-based quality control](read_quality_control.html) - Quality control of reads

The next step to get an overview is likely:

* [KRAKEN-based workflow](kraken.html) - if you don't know what your sample contains, and want to compare to known species
* [Alignment-based workflow](alignment.html) - if you have some clue what your sample contains
* [Informative KMER-based workflow](kmer.html) - if you don't know what your sample contains, and want to find novel genome compositions - Option #1
* [Count sketch KMER-based workflow](countsketch.html) - if you don't know what your sample contains, and want to find novel genome compositions - Option #2

Then there are several options on how to continue with deeper analysis:

* [Clustering](clustering.html) - to visualize your samples
* [De novo assembly](assembly.html)
* [Genome annotation](genome_annotation.html) - annotate your assemblies
* [Running software for each cell using MAP](map_scripts.html) - the general method for performing calculations on each cell

These may help you understand the design of Zorn and Bascet:

* [Overview of file formats](fileformats.html)
* [For developers](for_developers.html) - if you wish to modify Bascet
* [Example data](example_data.html) - you can run the workflow with some small datasets before you scale up


