<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>############################################### #' @return TODO #' @export LoadFragmentsAsObject &lt;- function(fragpath)  Index fragment file if needed. Only needed if you got the file externally and it has no index fragpath_index &lt;- paste(fragpath,".tbi",sep="") if(!file.exists(fragpath_index)) print("Indexing fragment file") system(paste("tabix -p vcf ",fragpath))   Figure out names of cells   Create a dummy assay stupidmat &lt;- matrix(0, nrow = 2, ncol=length(keep_cells)) rownames(stupidmat) &lt;- c("chr1:1-100", "chr1:200-300") colnames(stupidmat) &lt;- keep_cells chrom_assay &lt;- CreateChromatinAssay( counts = as.sparse(stupidmat), sep = c(":", "-"), fragments = fragpath, min.cells = 0, min.features = 0 ) adata &lt;- CreateSeuratObject( counts = chrom_assay, assay = "aligned" ) adata  — CountGrangeFeatures • Zorn</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="############################################### #' @return TODO #' @export LoadFragmentsAsObject &lt;- function(fragpath)  Index fragment file if needed. Only needed if you got the file externally and it has no index fragpath_index &lt;- paste(fragpath," .tbi if print fragment file system vcf figure out names of cells create a dummy assay stupidmat matrix nrow="2," ncol="length(keep_cells))" rownames c colnames keep_cells chrom_assay createchromatinassay counts="as.sparse(stupidmat)," sep='c(":",' fragments="fragpath," min.cells="0," min.features="0" adata createseuratobject countgrangefeatures><meta name="description" content="###############################################
#' @return TODO
#' @export
LoadFragmentsAsObject &amp;lt;- function(fragpath)

Index fragment file if needed. Only needed if you got the file externally and it has no indexfragpath_index &amp;lt;- paste(fragpath,&quot;.tbi&quot;,sep=&quot;&quot;)
if(!file.exists(fragpath_index))
print(&quot;Indexing fragment file&quot;)
system(paste(&quot;tabix -p vcf &quot;,fragpath))

Figure out names of cells
Create a dummy assaystupidmat &amp;lt;- matrix(0, nrow = 2, ncol=length(keep_cells))
rownames(stupidmat) &amp;lt;- c(&quot;chr1:1-100&quot;, &quot;chr1:200-300&quot;)
colnames(stupidmat) &amp;lt;- keep_cells
chrom_assay &amp;lt;- CreateChromatinAssay(
counts = as.sparse(stupidmat),
sep = c(&quot;:&quot;, &quot;-&quot;),
fragments = fragpath,
min.cells = 0,
min.features = 0
)
adata &amp;lt;- CreateSeuratObject(
counts = chrom_assay,
assay = &quot;aligned&quot;
)
adata
"><meta property="og:description" content="###############################################
#' @return TODO
#' @export
LoadFragmentsAsObject &amp;lt;- function(fragpath)

Index fragment file if needed. Only needed if you got the file externally and it has no indexfragpath_index &amp;lt;- paste(fragpath,&quot;.tbi&quot;,sep=&quot;&quot;)
if(!file.exists(fragpath_index))
print(&quot;Indexing fragment file&quot;)
system(paste(&quot;tabix -p vcf &quot;,fragpath))

Figure out names of cells
Create a dummy assaystupidmat &amp;lt;- matrix(0, nrow = 2, ncol=length(keep_cells))
rownames(stupidmat) &amp;lt;- c(&quot;chr1:1-100&quot;, &quot;chr1:200-300&quot;)
colnames(stupidmat) &amp;lt;- keep_cells
chrom_assay &amp;lt;- CreateChromatinAssay(
counts = as.sparse(stupidmat),
sep = c(&quot;:&quot;, &quot;-&quot;),
fragments = fragpath,
min.cells = 0,
min.features = 0
)
adata &amp;lt;- CreateSeuratObject(
counts = chrom_assay,
assay = &quot;aligned&quot;
)
adata
"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Zorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/get_started.html">Get started</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><h6 class="dropdown-header" data-toc-skip>Introductory Vignettes</h6></li>
    <li><a class="dropdown-item" href="../articles/debarcoding.html">Debarcoding</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Other stuff</h6></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../reference/faq.html">FAQ</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://henlab.org/"><span class="fa fa-home"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/henriksson-lab/zorn"><span class="fa fa-github"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>############################################### #' @return TODO #' @export LoadFragmentsAsObject &lt;- function(fragpath) <div class="section"> <h3 id="index-fragment-file-if-needed-only-needed-if-you-got-the-file-externally-and-it-has-no-index">Index fragment file if needed. Only needed if you got the file externally and it has no index<a class="anchor" aria-label="anchor" href="#index-fragment-file-if-needed-only-needed-if-you-got-the-file-externally-and-it-has-no-index"></a></h3> <p>fragpath_index &lt;- paste(fragpath,".tbi",sep="") if(!file.exists(fragpath_index)) print("Indexing fragment file") system(paste("tabix -p vcf ",fragpath))</p> </div> <div class="section"> <h3 id="figure-out-names-of-cells">Figure out names of cells<a class="anchor" aria-label="anchor" href="#figure-out-names-of-cells"></a></h3> </div> <div class="section"> <h3 id="create-a-dummy-assay">Create a dummy assay<a class="anchor" aria-label="anchor" href="#create-a-dummy-assay"></a></h3> <p>stupidmat &lt;- matrix(0, nrow = 2, ncol=length(keep_cells)) rownames(stupidmat) &lt;- c("chr1:1-100", "chr1:200-300") colnames(stupidmat) &lt;- keep_cells</p> <p>chrom_assay &lt;- CreateChromatinAssay( counts = as.sparse(stupidmat), sep = c(":", "-"), fragments = fragpath, min.cells = 0, min.features = 0 )</p> <p>adata &lt;- CreateSeuratObject( counts = chrom_assay, assay = "aligned" )</p> <p>adata</p> </div></h1>
      <small class="dont-index">Source: <a href="https://github.com/henriksson-lab/zorn/blob/main/R/refgenome.R" class="external-link"><code>R/refgenome.R</code></a></small>
      <div class="d-none name"><code>CountGrangeFeatures.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>###############################################
#' @return TODO
#' @export
LoadFragmentsAsObject &lt;- function(fragpath)
</p><div class="section">
<h3 id="index-fragment-file-if-needed-only-needed-if-you-got-the-file-externally-and-it-has-no-index-1">Index fragment file if needed. Only needed if you got the file externally and it has no index<a class="anchor" aria-label="anchor" href="#index-fragment-file-if-needed-only-needed-if-you-got-the-file-externally-and-it-has-no-index-1"></a></h3><p>fragpath_index &lt;- paste(fragpath,".tbi",sep="")
if(!file.exists(fragpath_index))
print("Indexing fragment file")
system(paste("tabix -p vcf ",fragpath))</p>
</div><div class="section">
<h3 id="figure-out-names-of-cells-1">Figure out names of cells<a class="anchor" aria-label="anchor" href="#figure-out-names-of-cells-1"></a></h3></div><div class="section">
<h3 id="create-a-dummy-assay-1">Create a dummy assay<a class="anchor" aria-label="anchor" href="#create-a-dummy-assay-1"></a></h3><p>stupidmat &lt;- matrix(0, nrow = 2, ncol=length(keep_cells))
rownames(stupidmat) &lt;- c("chr1:1-100", "chr1:200-300")
colnames(stupidmat) &lt;- keep_cells</p>
<p>chrom_assay &lt;- CreateChromatinAssay(
counts = as.sparse(stupidmat),
sep = c(":", "-"),
fragments = fragpath,
min.cells = 0,
min.features = 0
)</p>
<p>adata &lt;- CreateSeuratObject(
counts = chrom_assay,
assay = "aligned"
)</p>
<p>adata</p>
</div>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">CountGrangeFeatures</span><span class="op">(</span><span class="va">adata</span>, <span class="va">grange_gene</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>TODO</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Johan Henriksson, Laura Carroll, Hadrien Gourlé, Julian Dicken, Henriksson Lab, Carroll Lab and Collaborators.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

