<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index • Zorn</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Zorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/get_started.html">Get started</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><h6 class="dropdown-header" data-toc-skip>Introductory Vignettes</h6></li>
    <li><a class="dropdown-item" href="../articles/slurm.html">Using SLURM</a></li>
    <li><a class="dropdown-item" href="../articles/debarcoding.html">Debarcoding</a></li>
    <li><a class="dropdown-item" href="../articles/read_quality_control.html">Read-based quality control</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/kraken.html">KRAKEN-based workflow</a></li>
    <li><a class="dropdown-item" href="../articles/alignment.html">Alignment-based workflow</a></li>
    <li><a class="dropdown-item" href="../articles/kmer.html">Informative KMER-based workflow</a></li>
    <li><a class="dropdown-item" href="../articles/countsketch.html">Count sketch KMER-based workflow</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/clustering.html">Clustering</a></li>
    <li><a class="dropdown-item" href="../articles/assembly.html">De novo assembly</a></li>
    <li><a class="dropdown-item" href="../articles/genome_annotation.html">Genome annotation</a></li>
    <li><a class="dropdown-item" href="../articles/snp_analysis.html">SNP analysis</a></li>
    <li><a class="dropdown-item" href="../articles/map_scripts.html">Map scripts</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/fileformats.html">Overview of file formats</a></li>
    <li><a class="dropdown-item" href="../articles/bascet_files.html">Working with Bascet files</a></li>
    <li><a class="dropdown-item" href="../articles/for_developers.html">For developers</a></li>
    <li><a class="dropdown-item" href="../articles/example_data.html">Example data</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../articles/faq.html">FAQ</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://henlab.org/" aria-label="Lab website"><span class="fa fa-home"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/henriksson-lab/zorn" aria-label="GitHub repository"><span class="fa fa-github"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="AggregateMinhashes.html">AggregateMinhashes()</a></code>

        </dt>
        <dd>Aggregate frequency of minhashes across cells</dd>
      </dl><dl><dt>

          <code><a href="AtrandiBarcodeStats.html">AtrandiBarcodeStats()</a></code>

        </dt>
        <dd>Given a Bascet, produce a matrix showing for each combinatorial barcode, how many times it occurs across the cells. Presented as a 96-well plate matrix</dd>
      </dl><dl><dt>

          <code><a href="BarnyardPlotMatrix.html">BarnyardPlotMatrix()</a></code>

        </dt>
        <dd>Produce a matrix of Barnyard plots, i.e., counts for one species vs another, for all combinations of species.</dd>
      </dl><dl><dt>

          <code><a href="Bascet-class.html">Bascet-class</a></code>

        </dt>
        <dd>A bascet, along with all the shards</dd>
      </dl><dl><dt>

          <code><a href="BascetAggregateAMRfinder.html">BascetAggregateAMRfinder()</a></code>

        </dt>
        <dd>Aggregate data from AMRfinder This is a thin wrapper around BascetAggregateMap</dd>
      </dl><dl><dt>

          <code><a href="BascetAggregateAbricate.html">BascetAggregateAbricate()</a></code>

        </dt>
        <dd>Aggregate data from Abricate This is a thin wrapper around BascetAggregateMap</dd>
      </dl><dl><dt>

          <code><a href="BascetAggregateAriba.html">BascetAggregateAriba()</a></code>

        </dt>
        <dd>Aggregate data from Ariba This is a thin wrapper around BascetAggregateMap</dd>
      </dl><dl><dt>

          <code><a href="BascetAggregateFASTQC.html">BascetAggregateFASTQC()</a></code>

        </dt>
        <dd>Aggregate data from FASTQC This is a thin wrapper around BascetAggregateMap</dd>
      </dl><dl><dt>

          <code><a href="BascetAggregateGECCO.html">BascetAggregateGECCO()</a></code>

        </dt>
        <dd>Aggregate data from GECCO This is a thin wrapper around BascetAggregateMap</dd>
      </dl><dl><dt>

          <code><a href="BascetAggregateMap.html">BascetAggregateMap()</a></code>

        </dt>
        <dd>Aggregate data from previous Map call</dd>
      </dl><dl><dt>

          <code><a href="BascetAggregateQUAST.html">BascetAggregateQUAST()</a></code>

        </dt>
        <dd>Aggregate data from QUAST This is a thin wrapper around BascetAggregateMap</dd>
      </dl><dl><dt>

          <code><a href="BascetAlignToReference.html">BascetAlignToReference()</a></code>

        </dt>
        <dd>Align from FASTQ, generate sorted and indexed BAM file</dd>
      </dl><dl><dt>

          <code><a href="BascetAlignmentToBigwig.html">BascetAlignmentToBigwig()</a></code>

        </dt>
        <dd>Generate a bigwig out of all reads in a sorted BAM. Note that the caller is responsible for sorting the BAM first</dd>
      </dl><dl><dt>

          <code><a href="BascetBam2Fragments.html">BascetBam2Fragments()</a></code>

        </dt>
        <dd>Take aligned BAM file and produce Fragments.tsv.gz, compatible with Signac ATAC-seq style analysis</dd>
      </dl><dl><dt>

          <code><a href="BascetCacheComputation.html">BascetCacheComputation()</a></code>

        </dt>
        <dd>A wrapper to cache a computation. Put your function in as an argument, as R will only compute its value if needed. If the cache file exist, it will not be run again</dd>
      </dl><dl><dt>

          <code><a href="BascetCellNames.html">BascetCellNames()</a></code>

        </dt>
        <dd>Get list of cells in a Bascet</dd>
      </dl><dl><dt>

          <code><a href="BascetCellNames_withstreamer.html">BascetCellNames_withstreamer()</a></code>

        </dt>
        <dd>Get list of cells in a Bascet – streamer provided</dd>
      </dl><dl><dt>

          <code><a href="BascetComputeCountSketch.html">BascetComputeCountSketch()</a></code>

        </dt>
        <dd>Compute count sketch for each cell. This is a thin wrapper around BascetMapCell</dd>
      </dl><dl><dt>

          <code><a href="BascetComputeMinhash.html">BascetComputeMinhash()</a></code>

        </dt>
        <dd>Compute minhashes for each cell. This is a thin wrapper around BascetMapCell</dd>
      </dl><dl><dt>

          <code><a href="BascetCountChrom.html">BascetCountChrom()</a></code>

        </dt>
        <dd>From aligned BAM file, compute counts per chromosome</dd>
      </dl><dl><dt>

          <code><a href="BascetDumpContigs.html">BascetDumpContigs()</a></code>

        </dt>
        <dd>Store all contigs in an output directory, as cell_id.fa</dd>
      </dl><dl><dt>

          <code><a href="BascetFeaturiseKMC.html">BascetFeaturiseKMC()</a></code>

        </dt>
        <dd>Select kmers that appear useful for clustering</dd>
      </dl><dl><dt>

          <code><a href="BascetFilterAlignment.html">BascetFilterAlignment()</a></code>

        </dt>
        <dd>Filter an alignment (BAM-file).</dd>
      </dl><dl><dt>

          <code><a href="BascetGatherCountSketch.html">BascetGatherCountSketch()</a></code>

        </dt>
        <dd>Gather all count sketches into a single count sketch matrix</dd>
      </dl><dl><dt>

          <code><a href="BascetGetRaw.html">BascetGetRaw()</a></code>

        </dt>
        <dd>Generate BAM with barcodes from input raw FASTQ</dd>
      </dl><dl><dt>

          <code><a href="BascetIndexGenomeBWA.html">BascetIndexGenomeBWA()</a></code>

        </dt>
        <dd>Index a genome using BWA such that it can be used for alignment</dd>
      </dl><dl><dt>

          <code><a href="BascetIndexGenomeSTAR.html">BascetIndexGenomeSTAR()</a></code>

        </dt>
        <dd>Index a genome using STAR such that it can be used for alignment</dd>
      </dl><dl><dt>

          <code><a href="BascetInstance.html">BascetInstance()</a></code>

        </dt>
        <dd>Create a new bascet instance. For advanced users only</dd>
      </dl><dl><dt>

          <code><a href="BascetListFilesForCell.html">BascetListFilesForCell()</a></code>

        </dt>
        <dd>List files for a cell in a Bascet</dd>
      </dl><dl><dt>

          <code><a href="BascetLoadCountSketchMatrix.html">BascetLoadCountSketchMatrix()</a></code>

        </dt>
        <dd>Load count sketch matrix as Seurat object</dd>
      </dl><dl><dt>

          <code><a href="BascetMakeKrakenCountMatrix.html">BascetMakeKrakenCountMatrix()</a></code>

        </dt>
        <dd>Produce a count matrix of taxonomy IDs from KRAKEN output</dd>
      </dl><dl><dt>

          <code><a href="BascetMakeMinhashHistogram.html">BascetMakeMinhashHistogram()</a></code>

        </dt>
        <dd>Gather all minhashes into a single histogram file</dd>
      </dl><dl><dt>

          <code><a href="BascetMapCell.html">BascetMapCell()</a></code>

        </dt>
        <dd>Call a MAP function for all cells</dd>
      </dl><dl><dt>

          <code><a href="BascetMapCellAMRfinder.html">BascetMapCellAMRfinder()</a></code>

        </dt>
        <dd>Run AMRfinder on contigs of all cells. This is a thin wrapper around BascetMapCell</dd>
      </dl><dl><dt>

          <code><a href="BascetMapCellAbricate.html">BascetMapCellAbricate()</a></code>

        </dt>
        <dd>Run Abricate on contigs of all cells. This is a thin wrapper around BascetMapCell</dd>
      </dl><dl><dt>

          <code><a href="BascetMapCellAriba.html">BascetMapCellAriba()</a></code>

        </dt>
        <dd>Run Ariba on reads of all cells. This is a thin wrapper around BascetMapCell</dd>
      </dl><dl><dt>

          <code><a href="BascetMapCellBakta.html">BascetMapCellBakta()</a></code>

        </dt>
        <dd>Run Bakta on contigs of all cells. This is a thin wrapper around BascetMapCell</dd>
      </dl><dl><dt>

          <code><a href="BascetMapCellFASTQC.html">BascetMapCellFASTQC()</a></code>

        </dt>
        <dd>Run FASTQC on reads of all cells. This is a thin wrapper around BascetMapCell</dd>
      </dl><dl><dt>

          <code><a href="BascetMapCellGECCO.html">BascetMapCellGECCO()</a></code>

        </dt>
        <dd>Run GEECCO on contigs of all cells. This is a thin wrapper around BascetMapCell</dd>
      </dl><dl><dt>

          <code><a href="BascetMapCellQUAST.html">BascetMapCellQUAST()</a></code>

        </dt>
        <dd>Run QUAST on reads of all cells. This is a thin wrapper around BascetMapCell</dd>
      </dl><dl><dt>

          <code><a href="BascetMapTransform.html">BascetMapTransform()</a></code>

        </dt>
        <dd>Transform data</dd>
      </dl><dl><dt>

          <code><a href="BascetQueryFq.html">BascetQueryFq()</a></code>

        </dt>
        <dd>Build count table from FASTQ reads and a list of selected kmers</dd>
      </dl><dl><dt>

          <code><a href="BascetQueryKMC.html">BascetQueryKMC()</a></code>

        </dt>
        <dd>Build count table from kmer table and a list of selected kmers</dd>
      </dl><dl><dt>

          <code><a href="BascetReadFile.html">BascetReadFile()</a></code>

        </dt>
        <dd>Read one file from a Bascet</dd>
      </dl><dl><dt>

          <code><a href="BascetReadMinhashHistogram.html">BascetReadMinhashHistogram()</a></code>

        </dt>
        <dd>Read histogram of KMERs, the output of BascetMakeMinhashHistogram</dd>
      </dl><dl><dt>

          <code><a href="BascetRunCellSNP.html">BascetRunCellSNP()</a></code>

        </dt>
        <dd>Align from FASTQ, generate sorted and indexed BAM file</dd>
      </dl><dl><dt>

          <code><a href="BascetRunFASTP.html">BascetRunFASTP()</a></code>

        </dt>
        <dd>Run FASTP for each cell. Input must be in FASTQ file format</dd>
      </dl><dl><dt>

          <code><a href="BascetRunKraken.html">BascetRunKraken()</a></code>

        </dt>
        <dd>Run KRAKEN2 for each cell</dd>
      </dl><dl><dt>

          <code><a href="BascetShardify.html">BascetShardify()</a></code>

        </dt>
        <dd>Take debarcoded reads and split them into suitable numbers of shards.</dd>
      </dl><dl><dt>

          <code><a href="BascetShardify2.html">BascetShardify2()</a></code>

        </dt>
        <dd>Take debarcoded reads and split them into suitable numbers of shards.</dd>
      </dl><dl><dt>

          <code><a href="ChooseInformativeKMERs.html">ChooseInformativeKMERs()</a></code>

        </dt>
        <dd>Pick random KMERs from KMC3 database. The choice is among KMERs within a frequency range</dd>
      </dl><dl><dt>

          <code><a href="ChromToSpeciesCount.html">ChromToSpeciesCount()</a></code>

        </dt>
        <dd>Produce a count matrix on strain level</dd>
      </dl><dl><dt>

          <code><a href="CloseBascet.html">CloseBascet()</a></code>

        </dt>
        <dd>Close a Bascet file.</dd>
      </dl><dl><dt>

          <code><a href="CountDataFrameToSparseMatrix.html">CountDataFrameToSparseMatrix()</a></code>

        </dt>
        <dd>Count entries in long format data frame and return as a sparse matrix</dd>
      </dl><dl><dt>

          <code><a href="CountGrangeFeatures.html">CountGrangeFeatures()</a></code>

        </dt>
        <dd>Obtain a feature matrix (as seurat object) given an seurat object having Fragments associated</dd>
      </dl><dl><dt>

          <code><a href="CreateSeuratObjectWithReduction.html">CreateSeuratObjectWithReduction()</a></code>

        </dt>
        <dd>Create a seurat object from e.g. count sketch reduction</dd>
      </dl><dl><dt>

          <code><a href="DebarcodedKneePlot.html">DebarcodedKneePlot()</a></code>

        </dt>
        <dd>Produce summary kneeplot given debarcoded statistics</dd>
      </dl><dl><dt>

          <code><a href="DetectRawFileMeta.html">DetectRawFileMeta()</a></code>

        </dt>
        <dd>Detect metadata for raw input FASTQ files</dd>
      </dl><dl><dt>

          <code><a href="DownloadDatabaseAMRfinder.html">DownloadDatabaseAMRfinder()</a></code>

        </dt>
        <dd>Download a database for AMRfinder</dd>
      </dl><dl><dt>

          <code><a href="DownloadDatabaseAriba.html">DownloadDatabaseAriba()</a></code>

        </dt>
        <dd>Download database for Ariba</dd>
      </dl><dl><dt>

          <code><a href="DownloadDatabaseBakta.html">DownloadDatabaseBakta()</a></code>

        </dt>
        <dd>Download a database for Bakta</dd>
      </dl><dl><dt>

          <code><a href="FragmentCountsPerChrom.html">FragmentCountsPerChrom()</a></code>

        </dt>
        <dd>From a Signac chromatin assay with fragments, for each cell, count how many reads per chromosome</dd>
      </dl><dl><dt>

          <code><a href="FragmentCountsPerChromAssay.html">FragmentCountsPerChromAssay()</a></code>

        </dt>
        <dd>From a Signac chromatin assay with fragments, for each cell, count how many reads per chromosome. This function directly returns an assay that can be added to a Seurat multimodal object</dd>
      </dl><dl><dt>

          <code><a href="FragmentsToSignac.html">FragmentsToSignac()</a></code>

        </dt>
        <dd>From a fragments file, get a chromatin assay for Signac .......................... signac is dirt slow at counting! only 1 cpu used. take over the whole process</dd>
      </dl><dl><dt>

          <code><a href="GetBascetTempDir.html">GetBascetTempDir()</a></code>

        </dt>
        <dd>Get a temp directory to use; need to be created</dd>
      </dl><dl><dt>

          <code><a href="GetDefaultBascetInstance.html">GetDefaultBascetInstance()</a></code>

        </dt>
        <dd>Get Bascet instance from global variable</dd>
      </dl><dl><dt>

          <code><a href="GetFASTQCassembledDF.html">GetFASTQCassembledDF()</a></code>

        </dt>
        <dd>Get a data frame for one type of FASTQ statistics across across all cells</dd>
      </dl><dl><dt>

          <code><a href="GetFASTQCbasicStats.html">GetFASTQCbasicStats()</a></code>

        </dt>
        <dd>From aggregated FASTQC data, get basic statistics for overlay on UMAP etc</dd>
      </dl><dl><dt>

          <code><a href="GetFASTQCpassfailStats.html">GetFASTQCpassfailStats()</a></code>

        </dt>
        <dd>From aggregated FASTQC data, get overall pass-fail statistics for overlay on UMAP etc</dd>
      </dl><dl><dt>

          <code><a href="KmcChooseKmerFeatures.html">KmcChooseKmerFeatures()</a></code>

        </dt>
        <dd>Pick random KMERs from KMC3 database. The choice is among KMERs within a frequency range</dd>
      </dl><dl><dt>

          <code><a href="KmcGetHistogram.html">KmcGetHistogram()</a></code>

        </dt>
        <dd>Generate a histogram from a KMC3 database</dd>
      </dl><dl><dt>

          <code><a href="KmcGetKmers.html">KmcGetKmers()</a></code>

        </dt>
        <dd>Get KMERs from a KMC3 database</dd>
      </dl><dl><dt>

          <code><a href="KneeplotPerSpecies.html">KneeplotPerSpecies()</a></code>

        </dt>
        <dd>Produce a kneeplot</dd>
      </dl><dl><dt>

          <code><a href="KrakenFindConsensusTaxonomy.html">KrakenFindConsensusTaxonomy()</a></code>

        </dt>
        <dd>For a KRAKEN2 count matrix, return consensus taxID for each cell as metadata</dd>
      </dl><dl><dt>

          <code><a href="KrakenKneePlot.html">KrakenKneePlot()</a></code>

        </dt>
        <dd>Take a KRAKEN2 adata object and generate per-species kneeplots</dd>
      </dl><dl><dt>

          <code><a href="KrakenSpeciesDistribution.html">KrakenSpeciesDistribution()</a></code>

        </dt>
        <dd>Using a KRAKEN2 count matrix, produce a "kneeplot" of species</dd>
      </dl><dl><dt>

          <code><a href="ListDatabaseAbricate.html">ListDatabaseAbricate()</a></code>

        </dt>
        <dd>List installed databases available for Abricate</dd>
      </dl><dl><dt>

          <code><a href="LocalRunner.html">LocalRunner()</a></code>

        </dt>
        <dd>Create new local runner instance</dd>
      </dl><dl><dt>

          <code><a href="MapCellMultiListAsDataFrame.html">MapCellMultiListAsDataFrame()</a></code>

        </dt>
        <dd>Convenience function; alternative is to somehow implement as.data.frame</dd>
      </dl><dl><dt>

          <code><a href="MapListAsDataFrame.html">MapListAsDataFrame()</a></code>

        </dt>
        <dd>Convenience function; alternative is to somehow implement as.data.frame</dd>
      </dl><dl><dt>

          <code><a href="NoRunner.html">NoRunner()</a></code>

        </dt>
        <dd>Create new no-runner instance, used for debugging</dd>
      </dl><dl><dt>

          <code><a href="OpenBascet.html">OpenBascet()</a></code>

        </dt>
        <dd>Open a Bascet, prepare it for reading individual files</dd>
      </dl><dl><dt>

          <code><a href="PlotFASTQCadapterContent.html">PlotFASTQCadapterContent()</a></code>

        </dt>
        <dd>From aggregated FASTQC data, plot adapter content</dd>
      </dl><dl><dt>

          <code><a href="PlotHistogram.html">PlotHistogram()</a></code>

        </dt>
        <dd>Plot a histogram, loaded by ReadHistogram</dd>
      </dl><dl><dt>

          <code><a href="PlotJohnsonLindenstraussMinDim.html">PlotJohnsonLindenstraussMinDim()</a></code>

        </dt>
        <dd>Plot minimum number of dimensions needed to retain distance between samples</dd>
      </dl><dl><dt>

          <code><a href="PrepareSharding.html">PrepareSharding()</a></code>

        </dt>
        <dd>Prepare to shard reads by collecting statistics about each barcode, and filtering out cells with few reads</dd>
      </dl><dl><dt>

          <code><a href="ReadBascetCountMatrix.html">ReadBascetCountMatrix()</a></code>

        </dt>
        <dd>Read a count matrix as produced by Bascet (hdf5 format). This can be output from both BascetQueryFq and BascetCountChrom</dd>
      </dl><dl><dt>

          <code><a href="ReadCellSNPmatrix.html">ReadCellSNPmatrix()</a></code>

        </dt>
        <dd>Read a count matrix as produced by CellSNP, but as shards</dd>
      </dl><dl><dt>

          <code><a href="ReadHistogram.html">ReadHistogram()</a></code>

        </dt>
        <dd>Read the count histogram associated with a Bascet. Not all Bascets have one, but it is typically produced after debarcoding</dd>
      </dl><dl><dt>

          <code><a href="SetTaxonomyNamesFeatures.html">SetTaxonomyNamesFeatures()</a></code>

        </dt>
        <dd>Take a KRAKEN2 count matrix where the column is the taxonomyID. Convert to a matrix where the columns instead are the names of each taxonomy. Unused taxonomyID columns will not be kept</dd>
      </dl><dl><dt>

          <code><a href="ShowFASTQCforCell.html">ShowFASTQCforCell()</a></code>

        </dt>
        <dd>Show the FASTQC HTML report for a cell</dd>
      </dl><dl><dt>

          <code><a href="SlurmRunner.html">SlurmRunner()</a></code>

        </dt>
        <dd>Create a runner that submits jobs to SLURM</dd>
      </dl><dl><dt>

          <code><a href="TabixGetFragmentsSeqs.html">TabixGetFragmentsSeqs()</a></code>

        </dt>
        <dd>Using Tabix, get list of sequences in a fragment file</dd>
      </dl><dl><dt>

          <code><a href="TestBascetInstance.html">TestBascetInstance()</a></code>

        </dt>
        <dd>Check if a Bascet instance works</dd>
      </dl><dl><dt>

          <code><a href="aggr.abricate.html">aggr.abricate()</a></code>

        </dt>
        <dd>Callback function for aggregating ABRicate data for each cell. To be called from BascetAggregateMap</dd>
      </dl><dl><dt>

          <code><a href="aggr.amrfinder.html">aggr.amrfinder()</a></code>

        </dt>
        <dd>Callback function for aggregating ABRicate data for each cell. To be called from BascetAggregateMap</dd>
      </dl><dl><dt>

          <code><a href="aggr.ariba.html">aggr.ariba()</a></code>

        </dt>
        <dd>Callback function for aggregating ARIBA data for each cell. To be called from BascetAggregateMap</dd>
      </dl><dl><dt>

          <code><a href="aggr.fastqc.html">aggr.fastqc()</a></code>

        </dt>
        <dd>Callback function for aggregating FASTQC data for each cell. To be called from BascetAggregateMap</dd>
      </dl><dl><dt>

          <code><a href="aggr.gecco.html">aggr.gecco()</a></code>

        </dt>
        <dd>Callback function for aggregating GECCO data for each cell. To be called from BascetAggregateMap</dd>
      </dl><dl><dt>

          <code><a href="aggr.minhash.html">aggr.minhash()</a></code>

        </dt>
        <dd>Callback function for aggregating min-hashes for each cell. To be called from BascetAggregateMap</dd>
      </dl><dl><dt>

          <code><a href="aggr.quast.html">aggr.quast()</a></code>

        </dt>
        <dd>Callback function for aggregating QUAST data. To be called from BascetAggregateMap</dd>
      </dl><dl><dt>

          <code><a href="aggr.rawtext.html">aggr.rawtext()</a></code>

        </dt>
        <dd>Callback function for just getting raw file contents To be called from BascetAggregateMap</dd>
      </dl><dl><dt>

          <code><a href="bascetCheckOverwriteOutput.html">bascetCheckOverwriteOutput()</a></code>

        </dt>
        <dd>Detect metadata for raw input FASTQ files</dd>
      </dl><dl><dt>

          <code><a href="bascetInstance.default.html">bascetInstance.default</a></code>

        </dt>
        <dd>The default Bascet installation settings</dd>
      </dl><dl><dt>

          <code><a href="detectShardsForFile.html">detectShardsForFile()</a></code>

        </dt>
        <dd>Helper function: Figure out which shards belong together given root input name and extension i.e. root/name.##.ext</dd>
      </dl><dl><dt>

          <code><a href="extractstreamerExit.html">extractstreamerExit()</a></code>

        </dt>
        <dd>extract streamer: end an instance. the object should no longer be used after calling this function</dd>
      </dl><dl><dt>

          <code><a href="extractstreamerExtractTo.html">extractstreamerExtractTo()</a></code>

        </dt>
        <dd>extract streamer: extract to external file</dd>
      </dl><dl><dt>

          <code><a href="extractstreamerListCellsAnyFile.html">extractstreamerListCellsAnyFile()</a></code>

        </dt>
        <dd>extract streamer: list all cells in a given file</dd>
      </dl><dl><dt>

          <code><a href="extractstreamerLs.html">extractstreamerLs()</a></code>

        </dt>
        <dd>extract streamer: list all files in current zip-file</dd>
      </dl><dl><dt>

          <code><a href="extractstreamerOpen.html">extractstreamerOpen()</a></code>

        </dt>
        <dd>extract streamer: set which file is open</dd>
      </dl><dl><dt>

          <code><a href="extractstreamerReadNLines.html">extractstreamerReadNLines()</a></code>

        </dt>
        <dd>extract streamer: helper function to read N lines</dd>
      </dl><dl><dt>

          <code><a href="extractstreamerReadOneLine.html">extractstreamerReadOneLine()</a></code>

        </dt>
        <dd>extract streamer: helper function to read one line</dd>
      </dl><dl><dt>

          <code><a href="extractstreamerShowtext.html">extractstreamerShowtext()</a></code>

        </dt>
        <dd>extract streamer: get content of file, assumed to be text (or this function crashes)</dd>
      </dl><dl><dt>

          <code><a href="extractstreamerStart.html">extractstreamerStart()</a></code>

        </dt>
        <dd>extract streamer: create an instance</dd>
      </dl><dl><dt>

          <code><a href="getBascetDockerImage.html">getBascetDockerImage()</a></code>

        </dt>
        <dd>Get and install a Bascet docker image.</dd>
      </dl><dl><dt>

          <code><a href="getBascetSingularityImage.html">getBascetSingularityImage()</a></code>

        </dt>
        <dd>Get a Bascet image (singularity or docker). It will be cached in the provided directory to avoid downloading it all the time</dd>
      </dl><dl><dt>

          <code><a href="getFastqR2fromR1.html">getFastqR2fromR1()</a></code>

        </dt>
        <dd>Get corresponding R2 file. Assumes that the input file is R1</dd>
      </dl><dl><dt>

          <code><a href="isBamPairedAlignment.html">isBamPairedAlignment()</a></code>

        </dt>
        <dd>Figure out if a BAM-file is a paired alignment or not</dd>
      </dl><dl><dt>

          <code><a href="isFastq.html">isFastq()</a></code>

        </dt>
        <dd>Check if a file is a FASTQ file</dd>
      </dl><dl><dt>

          <code><a href="isPairedFastq.html">isPairedFastq()</a></code>

        </dt>
        <dd>Check if a file is a paired FASTQ file</dd>
      </dl><dl><dt>

          <code><a href="lseq.html">lseq()</a></code>

        </dt>
        <dd>logarithmic spaced sequence, from emdbook library</dd>
      </dl><dl><dt>

          <code><a href="makeOutputShardNames.html">makeOutputShardNames()</a></code>

        </dt>
        <dd>Helper function: Generate suitable output filenames according to shard system i.e. root/name.##.ext</dd>
      </dl><dl><dt>

          <code><a href="new_no_job.html">new_no_job()</a></code>

        </dt>
        <dd>Create an empty job. It is considered to have terminated from the start</dd>
      </dl><dl><dt>

          <code><a href="removeBascetDockerImage.html">removeBascetDockerImage()</a></code>

        </dt>
        <dd>Remove current Bascet docker image</dd>
      </dl><dl><dt>

          <code><a href="shellscriptCancelJobIfFileExists.html">shellscriptCancelJobIfFileExists()</a></code>

        </dt>
        <dd>Create a piece of script to exit a job early if file exists</dd>
      </dl><dl><dt>

          <code><a href="shellscriptMakeCommalist.html">shellscriptMakeCommalist()</a></code>

        </dt>
        <dd>from a,b to "a,b"</dd>
      </dl><dl><dt>

          <code><a href="shellscript_make_bash_array.html">shellscript_make_bash_array()</a></code>

        </dt>
        <dd>Helper function: Create array of values in bash scripts</dd>
      </dl><dl><dt>

          <code><a href="template_BascetFunction.html">template_BascetFunction()</a></code>

        </dt>
        <dd>Function template, where basic parameter documentation can be obtained from</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Johan Henriksson, Laura Carroll, Hadrien Gourlé, Julian Dicken, Henriksson Lab, Carroll Lab and Collaborators.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

