<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Alignment-based workflow • Zorn</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Alignment-based workflow">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Zorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/get_started.html">Get started</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><h6 class="dropdown-header" data-toc-skip>Introductory Vignettes</h6></li>
    <li><a class="dropdown-item" href="../articles/slurm.html">Using SLURM</a></li>
    <li><a class="dropdown-item" href="../articles/debarcoding.html">Debarcoding</a></li>
    <li><a class="dropdown-item" href="../articles/read_quality_control.html">Read-based quality control</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/kraken.html">KRAKEN-based workflow</a></li>
    <li><a class="dropdown-item" href="../articles/alignment.html">Alignment-based workflow</a></li>
    <li><a class="dropdown-item" href="../articles/kmer.html">Informative KMER-based workflow</a></li>
    <li><a class="dropdown-item" href="../articles/countsketch.html">Count sketch KMER-based workflow</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/clustering.html">Clustering</a></li>
    <li><a class="dropdown-item" href="../articles/assembly.html">De novo assembly</a></li>
    <li><a class="dropdown-item" href="../articles/genome_annotation.html">Genome annotation</a></li>
    <li><a class="dropdown-item" href="../articles/snp_analysis.html">SNP analysis</a></li>
    <li><a class="dropdown-item" href="../articles/map_scripts.html">Map scripts</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/fileformats.html">Overview of file formats</a></li>
    <li><a class="dropdown-item" href="../articles/bascet_files.html">Working with Bascet files</a></li>
    <li><a class="dropdown-item" href="../articles/for_developers.html">For developers</a></li>
    <li><a class="dropdown-item" href="../articles/example_data.html">Example data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../articles/faq.html">FAQ</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://henlab.org/" aria-label="Lab website"><span class="fa fa-home"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/henriksson-lab/zorn" aria-label="GitHub repository"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Alignment-based workflow</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/henriksson-lab/zorn/blob/main/vignettes/alignment.Rmd" class="external-link"><code>vignettes/alignment.Rmd</code></a></small>
      <div class="d-none name"><code>alignment.Rmd</code></div>
    </div>

    
    
<p>First set up your Zorn/Bascet workdirectory as before. If you wish to
run these steps on a SLURM cluster, see separate vignette and adapt
accordingly.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/henriksson-lab/zorn" class="external-link">Zorn</a></span><span class="op">)</span></span>
<span><span class="va">bascet_runner.default</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LocalRunner.html">LocalRunner</a></span><span class="op">(</span>direct <span class="op">=</span> <span class="cn">TRUE</span>, showScript<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">bascetRoot</span> <span class="op">&lt;-</span> <span class="st">"/home/yours/an_empty_workdirectory"</span></span></code></pre></div>
<div class="section level2">
<h2 id="alignment">Alignment<a class="anchor" aria-label="anchor" href="#alignment"></a>
</h2>
<p>We will use BWA for alignment, which wants the data in FASTQ format.
Here we assume that you use a single-cell WGS chemistry that produces
paired reads, and direct Zorn to produce R1/R2 FASTQ files (only R1
specified). We name the output Bascets “asfq”, taking as input the
typical sharded reads:</p>
<p><a href="slurm.html">(SLURM-compatible step)</a></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Get reads in fastq format</span></span>
<span><span class="fu"><a href="../reference/BascetMapTransform.html">BascetMapTransform</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>,</span>
<span>  inputName<span class="op">=</span><span class="st">"filtered"</span>,   <span class="co">#default; can omit</span></span>
<span>  outputName<span class="op">=</span><span class="st">"asfq"</span>,  <span class="co">###name parameter</span></span>
<span>  out_format<span class="op">=</span><span class="st">"R1.fq.gz"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Before alignment, you need to prepare your reference by indexing it.
You can do this using “bwa index” in the terminal; but you can also do
it via Zorn. This assumes that you have a FASTA-file named all.fa (can
also be gzipped):</p>
<p><a href="slurm.html">(SLURM-compatible step)</a></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Build a reference</span></span>
<span><span class="fu"><a href="../reference/BascetIndexGenomeBWA.html">BascetIndexGenomeBWA</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>,</span>
<span>  <span class="st">"/your_reference/all.fa"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>You can now proceed with alignment using BWA. By default, this
command outputs both unsorted and sorted BAM-files.</p>
<ul>
<li>The sorted BAM-files are required for counting of reads in genomic
regions, i.e., WGS species counting and RNA-seq</li>
<li>The unsorted BAM-files are required for filtering of host reads</li>
</ul>
<p><a href="slurm.html">(SLURM-compatible step)</a></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Perform alignment</span></span>
<span><span class="fu"><a href="../reference/BascetAlignToReference.html">BascetAlignToReference</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>,</span>
<span>  useReference<span class="op">=</span><span class="st">"/your_reference/all.fa"</span>,</span>
<span>  numLocalThreads<span class="op">=</span><span class="fl">10</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>TODO: Filtering</p>
</div>
<div class="section level2">
<h2 id="host-filtering-workflow">Host filtering workflow<a class="anchor" aria-label="anchor" href="#host-filtering-workflow"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/BascetFilterAlignment.html">BascetFilterAlignment</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>,</span>
<span>  outputName<span class="op">=</span><span class="st">"filter_paired"</span>,</span>
<span>  keep_mapped<span class="op">=</span><span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="option-1-counting-with-signac">Option #1: Counting with Signac<a class="anchor" aria-label="anchor" href="#option-1-counting-with-signac"></a>
</h2>
<p>If you want to analyze where reads mapped, either for RNA-seq
counting for genes, or more generally, what the count is within a
genomic region, you can use Signac for great flexibility in this task.
Signac is designed to analyze single-cell ATAC-seq data, where the reads
are stored in a specialized BED-file. You can use the command below to
generate this file.</p>
<p>Note one caveat: depending on chemistry, this workflow may not handle
UMIs, and just report raw read counts. However, for quick and dirty
operations this is typically enough.</p>
<p><a href="slurm.html">(SLURM-compatible step)</a></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Generate fragments BED file suited for quantifying</span></span>
<span><span class="co">### reads/chromosome using Signac later </span></span>
<span><span class="fu"><a href="../reference/BascetBam2Fragments.html">BascetBam2Fragments</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="option-2-counting-chromosomes-species">Option #2: Counting chromosomes (species)<a class="anchor" aria-label="anchor" href="#option-2-counting-chromosomes-species"></a>
</h2>
<p>A simplified scenario is when you just want to get the number of
reads per chromosome. These counts can later be summarized on the levels
of genomes (summing up across chromosomes). We use this to for example
compare the alignment to an expected mock community, and thus just want
to see how species mix together. The following command generates counts
files:</p>
<p><a href="slurm.html">(SLURM-compatible step)</a></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Count reads per chromosome</span></span>
<span><span class="fu"><a href="../reference/BascetCountChrom.html">BascetCountChrom</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The output format is one binary HDF5 file for each shard, roughly in
the <a href="https://anndata.readthedocs.io/en/stable/fileformat-prose.html" class="external-link">Anndata
file format</a>. To load these, use the following command that loads the
files and concatenates them into a single matrix. It can then be loaded
into Seurat:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">cnt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ReadBascetCountMatrix.html">ReadBascetCountMatrix</a></span><span class="op">(</span><span class="va">bascetRoot</span>,<span class="st">"chromcount"</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span></span>
<span>  counts <span class="op">=</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/CreateAssayObject.html" class="external-link">CreateAssayObject</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">cnt</span><span class="op">$</span><span class="va">X</span><span class="op">)</span><span class="op">)</span>, <span class="co">## according to anndata standard, there should be a transpose here</span></span>
<span>  assay <span class="op">=</span> <span class="st">"chrom_cnt"</span></span>
<span><span class="op">)</span> </span></code></pre></div>
<p>Most likely, some cells will have to be discarded. Initially we
recommend that you keep the cutoff low until you are better informed.
You can then apply the cutoff as follows:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keep_cells</span> <span class="op">&lt;-</span> <span class="va">adata</span><span class="op">$</span><span class="va">nCount_chrom_cnt</span> <span class="op">&gt;</span> <span class="fl">10000</span> <span class="co">#10k reads</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">keep_cells</span><span class="op">)</span>                              <span class="co">#See how many cells pass this cutoff</span></span>
<span><span class="va">adata</span> <span class="op">&lt;-</span> <span class="va">adata</span><span class="op">[</span>,<span class="va">keep_cells</span><span class="op">]</span>                  <span class="co">#Reduce to sensible number</span></span></code></pre></div>
<p>If you have a list of which chromosomes belong together, then you can
furthermore sum them up to get species-level counts. mapSeq2strain
should contain two columns: (todo) (TODO filter after instead)</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">adata</span><span class="op">[[</span><span class="st">"species_cnt"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ChromToSpeciesCount.html">ChromToSpeciesCount</a></span><span class="op">(</span><span class="va">adata</span>, <span class="va">mapSeq2strain</span><span class="op">)</span>  </span>
<span></span>
<span><span class="co">#Optional: Figure out which species has most reads in which cell</span></span>
<span><span class="va">cnt</span> <span class="op">&lt;-</span> <span class="va">adata</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">species_cnt</span><span class="op">$</span><span class="va">counts</span></span>
<span><span class="va">adata</span><span class="op">$</span><span class="va">species_aln</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cnt</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">cnt</span>, <span class="fl">2</span>, <span class="va">which.max</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>Knowing the species, you can generate a per-species kneeplot as
follows:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"species_cnt"</span></span>
<span><span class="fu"><a href="../reference/KneeplotPerSpecies.html">KneeplotPerSpecies</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span></span></code></pre></div>
<p>You can now proceed to generate a dimensional reduction in the form
of a UMAP. See the tutorials for <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a> and <a href="https://stuartlab.org/signac/" class="external-link">Signac</a> for details. We
summarize the required command below, where you can apply either an
RNA-seq or ATAC-seq dimensional reduction depending on what you think is
appropriate.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co">#ATAC-seq style analysis</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stuart-lab/signac" class="external-link">Signac</a></span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/RunTFIDF.html" class="external-link">RunTFIDF</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/FindTopFeatures.html" class="external-link">FindTopFeatures</a></span><span class="op">(</span><span class="va">adata</span>, min.cutoff <span class="op">=</span> <span class="st">'q0'</span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/RunSVD.html" class="external-link">RunSVD</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://stuartlab.org/signac/reference/DepthCor.html" class="external-link">DepthCor</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">adata</span>, reduction <span class="op">=</span> <span class="st">'lsi'</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span>, reduction.name <span class="op">=</span> <span class="st">"chrom_umap"</span><span class="op">)</span>  </span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="co">#RNA-seq style analysis</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">adata</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span>, nfeatures <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">adata</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">adata</span>, features <span class="op">=</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">adata</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">adata</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span>, reduction.name <span class="op">=</span> <span class="st">"chrom_umap"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Finally, you can plot your UMAP. If you also extracted the dominant
species column in your matrix, you can visualize it on top!</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">adata</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, group.by <span class="op">=</span> <span class="st">"species_aln"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">xlab</span><span class="op">(</span><span class="st">"BWA1"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ylab</span><span class="op">(</span><span class="st">"BWA2"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="option-3-rna-seq-analysis-feature-counting">Option #3: RNA-seq analysis (feature counting)<a class="anchor" aria-label="anchor" href="#option-3-rna-seq-analysis-feature-counting"></a>
</h2>
<p>For protocols such as RNA-seq, the aim is to get the number of reads
per gene, per cell. After alignment and sorting of reads based on genome
coordinate (default setting), you can overlay the reads with gene
coordinates from a GTF or GFF file (both are supported). The following
command performs this operation:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/BascetCountFeature.html">BascetCountFeature</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>,</span>
<span>  gffFile <span class="op">=</span> <span class="st">"/my/genome.gtf.gz"</span>,</span>
<span>  useFeature <span class="op">=</span> <span class="st">"gene"</span>,     <span class="co">#Default</span></span>
<span>  attrGeneId <span class="op">=</span> <span class="st">"gene_id"</span>,  <span class="co">#Default</span></span>
<span>  attrGeneName <span class="op">=</span> <span class="st">"name"</span>    <span class="co">#Default</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The ID and name of genes are extracted from the GFF attributes
column. Not all files have both name and ID; the ID will be used as name
if nothing specified. By default, the rows having “gene” as the feature
description will</p>
<p>After the features (genes) have been counted, you can load them into
Seurat:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">cnt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ReadBascetCountMatrix.html">ReadBascetCountMatrix</a></span><span class="op">(</span><span class="va">bascetRoot</span>,<span class="st">"featurecount"</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span></span>
<span>  counts <span class="op">=</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/CreateAssayObject.html" class="external-link">CreateAssayObject</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">cnt</span><span class="op">$</span><span class="va">X</span><span class="op">)</span><span class="op">)</span> <span class="co">## according to anndata standard, there should be a transpose here</span></span>
<span><span class="op">)</span> </span></code></pre></div>
<p>From here, you can follow an <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial" class="external-link">RNA-seq
tutorial for Seurat</a>. But for the impatient, this is some minimal
example code to get your first UMAP:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span></span>
<span><span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">adata</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span>, nfeatures <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">adata</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">adata</span>, features <span class="op">=</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">adata</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">adata</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">adata</span>, resolution <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">adata</span>, dims<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Johan Henriksson, Laura Carroll, Hadrien Gourlé, Julian Dicken, Henriksson Lab, Carroll Lab and Collaborators.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
