<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Alignment-based workflow • Zorn</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Alignment-based workflow">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Zorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/get_started.html">Get started</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><h6 class="dropdown-header" data-toc-skip>Introductory Vignettes</h6></li>
    <li><a class="dropdown-item" href="../articles/debarcoding.html">Debarcoding</a></li>
    <li><a class="dropdown-item" href="../articles/kraken.html">KRAKEN-based workflow</a></li>
    <li><a class="dropdown-item" href="../articles/alignment.html">Alignment-based workflow</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/fileformats.html">Overview of file formats</a></li>
    <li><a class="dropdown-item" href="../articles/slurm.html">Using SLURM</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../articles/faq.html">FAQ</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://henlab.org/"><span class="fa fa-home"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/henriksson-lab/zorn"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Alignment-based workflow</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/henriksson-lab/zorn/blob/main/vignettes/alignment.Rmd" class="external-link"><code>vignettes/alignment.Rmd</code></a></small>
      <div class="d-none name"><code>alignment.Rmd</code></div>
    </div>

    
    
<p>First set up your Zorn/Bascet workdirectory as before. If you wish to
run these steps on a SLURM cluster, see separate vignette and adapt
accordingly.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/henriksson-lab/zorn" class="external-link">Zorn</a></span><span class="op">)</span></span>
<span><span class="va">bascet_runner</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LocalRunner.html">LocalRunner</a></span><span class="op">(</span>direct <span class="op">=</span> <span class="cn">TRUE</span>, show_script<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">bascetRoot</span> <span class="op">=</span> <span class="st">"/home/yours/an_empty_workdirectory"</span></span></code></pre></div>
<p>We will use BWA for alignment, which wants the data in FASTQ format.
Here we assume that you use a single-cell WGS chemistry that produces
paired reads, and direct Zorn to produce R1/R2 FASTQ files (only R1
specified). We name the output Bascets “asfq”, taking as input the
typical sharded reads:</p>
<p>(SLURM-compatible)</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Get reads in fastq format</span></span>
<span><span class="fu"><a href="../reference/BascetMapTransform.html">BascetMapTransform</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>,</span>
<span>  <span class="st">"filtered"</span>,   <span class="co">#default; can omit</span></span>
<span>  <span class="st">"asfq"</span>,  <span class="co">###name parameter</span></span>
<span>  out_format<span class="op">=</span><span class="st">"R1.fq.gz"</span>,</span>
<span>  runner<span class="op">=</span><span class="va">bascet_runner</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Before alignment, you need to prepare your reference by indexing it.
You can do this using “bwa index” in the terminal; but you can also do
it via Zorn. This assumes that you have a FASTA-file named all.fa (can
also be gzipped):</p>
<p>(SLURM-compatible)</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Build a reference</span></span>
<span><span class="fu"><a href="../reference/BascetIndexGenomeBWA.html">BascetIndexGenomeBWA</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>, </span>
<span>  <span class="st">"/your_reference/all.fa"</span>, </span>
<span>  runner<span class="op">=</span><span class="va">bascet_runner</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>You can now proceed with alignment using BWA. By default, this
command outputs both unsorted and sorted BAM-files.</p>
<ul>
<li>The sorted BAM-files are required for counting of reads in genomic
regions, i.e., WGS species counting and RNA-seq</li>
<li>The unsorted BAM-files are required for filtering of host reads</li>
</ul>
<p>(SLURM-compatible)</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Perform alignment</span></span>
<span><span class="fu"><a href="../reference/BascetAlignToReference.html">BascetAlignToReference</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>,</span>
<span>  useReference<span class="op">=</span><span class="st">"/your_reference/all.fa"</span>,</span>
<span>  numLocalThreads<span class="op">=</span><span class="fl">10</span>,</span>
<span>  runner<span class="op">=</span><span class="va">bascet_runner</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>TODO: Filtering</p>
<p>==================== Host filtering workflow ==============</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/BascetFilterAlignment.html">BascetFilterAlignment</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>,</span>
<span>  outputName<span class="op">=</span><span class="st">"filter_paired"</span>,</span>
<span>  keep_mapped<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>  runner<span class="op">=</span><span class="va">bascet_runner</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>==================== Counting workflow ==============</p>
<p>If you want to analyze where reads mapped, either for RNA-seq
counting for genes, or more generally, what the count is within a
genomic region, you can use Signac for great flexibility in this task.
Signac is designed to analyze single-cell ATAC-seq data, where the reads
are stored in a specialized BED-file. You can use the command below to
generate this file.</p>
<p>Note one caveat: depending on chemistry, this workflow may not handle
UMIs, and just report raw read counts. However, for quick and dirty
operations this is typically enough.</p>
<p>(SLURM-compatible)</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Generate fragments BED file suited for quantifying reads/chromosome using Signac later </span></span>
<span><span class="fu"><a href="../reference/BascetBam2Fragments.html">BascetBam2Fragments</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>,</span>
<span>  runner<span class="op">=</span><span class="va">bascet_runner</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>If you just want to get the number of reads per chromosome (i.e. to
compare a metagenomic sample to a number of defined strains), then we
provide a simpler counter. The output is a small matrix that you can
load with Seurat:</p>
<p>(SLURM-compatible)</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Count reads per chromosome</span></span>
<span><span class="fu"><a href="../reference/BascetCountChrom.html">BascetCountChrom</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>, </span>
<span>  runner<span class="op">=</span><span class="va">bascet_runner</span></span>
<span><span class="op">)</span></span></code></pre></div>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Johan Henriksson, Laura Carroll, Hadrien Gourlé, Julian Dicken, Henriksson Lab, Carroll Lab and Collaborators.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
