<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Working with Bascet-FASTQ • Zorn</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Working with Bascet-FASTQ">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Zorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/get_started.html">Get started</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><h6 class="dropdown-header" data-toc-skip>Introductory Vignettes</h6></li>
    <li><a class="dropdown-item" href="../articles/slurm.html">Using SLURM</a></li>
    <li><a class="dropdown-item" href="../articles/debarcoding.html">Debarcoding</a></li>
    <li><a class="dropdown-item" href="../articles/read_quality_control.html">Read-based quality control</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/kraken.html">KRAKEN-based workflow</a></li>
    <li><a class="dropdown-item" href="../articles/alignment.html">Alignment-based workflow</a></li>
    <li><a class="dropdown-item" href="../articles/kmer.html">Informative KMER-based workflow</a></li>
    <li><a class="dropdown-item" href="../articles/countsketch.html">Count sketch KMER-based workflow</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/assembly.html">De novo assembly</a></li>
    <li><a class="dropdown-item" href="../articles/genome_annotation.html">Genome annotation</a></li>
    <li><a class="dropdown-item" href="../articles/snp_analysis.html">SNP analysis</a></li>
    <li><a class="dropdown-item" href="../articles/map_scripts.html">Map scripts</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/fileformats.html">Overview of file formats</a></li>
    <li><a class="dropdown-item" href="../articles/bascet_zip.html">The types of Bascet-ZIP files</a></li>
    <li><a class="dropdown-item" href="../articles/bascet_fastq.html">Working with Bascet-FASTQ</a></li>
    <li><a class="dropdown-item" href="../articles/for_developers.html">For developers</a></li>
    <li><a class="dropdown-item" href="../articles/example_data.html">Example data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../articles/faq.html">FAQ</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://henlab.org/" aria-label="Lab website"><span class="fa fa-home"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/henriksson-lab/zorn" aria-label="GitHub repository"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Working with Bascet-FASTQ</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/henriksson-lab/zorn/blob/main/vignettes/bascet_fastq.Rmd" class="external-link"><code>vignettes/bascet_fastq.Rmd</code></a></small>
      <div class="d-none name"><code>bascet_fastq.Rmd</code></div>
    </div>

    
    
<p>Advanced users may wish to process the reads using custom software.
As such software typically ingests FASTQ-files, we provide Bascet-FASTQ.
These files are regular <a href="fileformats.html">(pairwise FASTQ-files
with a naming scheme)</a>.</p>
<p>There are a few things you can do with these: * Special trimming (but
Bascet already performs rudimentary trimming) * Use alternative aligners
(the output BAM files retains read names and can thus be ingested by
Bascet again) * Extract special sequences for separate analysis
(perturb-seq sgRNAs; CITE-seq; etc) * Deposit the FASTQ into archives
(but read separate section!)</p>
<p>We assume that you have set up for Instance and Runner for the
following commands to work.</p>
<div class="section level2">
<h2 id="raw-fastq-depositing">Raw FASTQ depositing<a class="anchor" aria-label="anchor" href="#raw-fastq-depositing"></a>
</h2>
<p>Bascet-FASTQ is possibly suitable for depositing raw data. The data
will take less space as trimming and barcode identification has already
been performed. The sorting also improves compression ratio. The
presorting further reduces work for reproducing the analysis.</p>
<p><em>However, there is a major caveat:</em> FASTQ-files submitted to
SRA will have the names of reads removed in their “light” format (i.e.,
which cell the read belongs to). It will still be possible to get the
fully raw files using the cold storage retrieval option, but users may
have to pay for this, and download takes longer times.</p>
<p>We are working toward a solution, but users are currently better off
submitting raw data to Zenodo or other general depository. In such a
case, you might as well submit the native TIRP files, but we currently
do not make any recommendation.</p>
</div>
<div class="section level2">
<h2 id="conversion-to-bascet-fastq">Conversion to Bascet-FASTQ<a class="anchor" aria-label="anchor" href="#conversion-to-bascet-fastq"></a>
</h2>
<p>The following command converts any Bascet file containing reads to
Bascet-FASTQ:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Get reads in FASTQ format</span></span>
<span><span class="fu"><a href="../reference/BascetMapTransform.html">BascetMapTransform</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>,</span>
<span>  inputName<span class="op">=</span><span class="st">"filtered"</span>,</span>
<span>  outputName<span class="op">=</span><span class="st">"asfq"</span>,</span>
<span>  outFormat<span class="op">=</span><span class="st">"R1.fq.gz"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="trimming-with-fastp">Trimming with FASTP<a class="anchor" aria-label="anchor" href="#trimming-with-fastp"></a>
</h2>
<p><a href="https://github.com/OpenGene/fastp" class="external-link">(FASTP)</a> is a fast
trimmer. We however don’t recommend it for de novo assembly as it seems
to leave too many adapters - <em>this is why we already perform our own
trimming during GetRaw</em>. In either case, FASTP can be run directly
through Zorn. If you wish to use other software, you can apply it in any
other way to the FASTQ files.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Get reads in fastq format</span></span>
<span><span class="fu"><a href="../reference/BascetRunFASTP.html">BascetRunFASTP</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>,</span>
<span>  numLocalThreads<span class="op">=</span><span class="fl">10</span>,</span>
<span>  inputName<span class="op">=</span><span class="st">"asfq"</span>,</span>
<span>  outputName<span class="op">=</span><span class="st">"fastp"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="conversion-from-bascet-fastq-to-tirp">Conversion from Bascet-FASTQ to TIRP<a class="anchor" aria-label="anchor" href="#conversion-from-bascet-fastq-to-tirp"></a>
</h2>
<p>We recommend keeping reads in TIRP files as this is what commands are
optimized for; and TIRP offers direct extraction of reads from a cell,
via indexing, unlike FASTQ. The following command performs the
conversion</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Convert FASTQ to TIRP</span></span>
<span><span class="fu"><a href="../reference/BascetMapTransform.html">BascetMapTransform</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>,</span>
<span>  <span class="st">"fastp"</span>,</span>
<span>  <span class="st">"new_filtered"</span>,</span>
<span>  outFormat<span class="op">=</span><span class="st">"tirp.gz"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>You now have a TIRP file again, equivalent to the first shardified
output. !!! Don’t forget to specify inputName=“new_filtered” to later
commands, as the default is to use “filtered”</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Johan Henriksson, Laura Carroll, Hadrien Gourlé, Julian Dicken, HenLab, CompMicroLab, and Collaborators.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
