<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Count sketch KMER-based workflow • Zorn</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Count sketch KMER-based workflow">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Zorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/get_started.html">Get started</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><h6 class="dropdown-header" data-toc-skip>Introductory Vignettes</h6></li>
    <li><a class="dropdown-item" href="../articles/slurm.html">Using SLURM</a></li>
    <li><a class="dropdown-item" href="../articles/debarcoding.html">Debarcoding</a></li>
    <li><a class="dropdown-item" href="../articles/read_quality_control.html">Read-based quality control</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/kraken.html">KRAKEN-based workflow</a></li>
    <li><a class="dropdown-item" href="../articles/alignment.html">Alignment-based workflow</a></li>
    <li><a class="dropdown-item" href="../articles/kmer.html">Informative KMER-based workflow</a></li>
    <li><a class="dropdown-item" href="../articles/countsketch.html">Count sketch KMER-based workflow</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/assembly.html">De novo assembly</a></li>
    <li><a class="dropdown-item" href="../articles/genome_annotation.html">Genome annotation</a></li>
    <li><a class="dropdown-item" href="../articles/snp_analysis.html">SNP analysis</a></li>
    <li><a class="dropdown-item" href="../articles/map_scripts.html">Map scripts</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/fileformats.html">Overview of file formats</a></li>
    <li><a class="dropdown-item" href="../articles/bascet_zip.html">The types of Bascet-ZIP files</a></li>
    <li><a class="dropdown-item" href="../articles/bascet_fastq.html">Working with Bascet-FASTQ</a></li>
    <li><a class="dropdown-item" href="../articles/for_developers.html">For developers</a></li>
    <li><a class="dropdown-item" href="../articles/example_data.html">Example data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../articles/faq.html">FAQ</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://henlab.org/" aria-label="Lab website"><span class="fa fa-home"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/henriksson-lab/zorn" aria-label="GitHub repository"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Count sketch KMER-based workflow</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/henriksson-lab/zorn/blob/main/vignettes/countsketch.Rmd" class="external-link"><code>vignettes/countsketch.Rmd</code></a></small>
      <div class="d-none name"><code>countsketch.Rmd</code></div>
    </div>

    
    
<p>First set up your Zorn/Bascet work directory as before. If you wish
to run these steps on a SLURM cluster, see separate vignette and adapt
accordingly.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/henriksson-lab/zorn" class="external-link">Zorn</a></span><span class="op">)</span></span>
<span><span class="va">bascetRunner.default</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LocalRunner.html">LocalRunner</a></span><span class="op">(</span>direct <span class="op">=</span> <span class="cn">TRUE</span>, showScript<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">bascetInstance.default</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getBascetSingularityImage.html">getBascetSingularityImage</a></span><span class="op">(</span>storeAt<span class="op">=</span><span class="st">"~/"</span><span class="op">)</span> <span class="co">#Assuming Linux</span></span>
<span><span class="va">bascetRoot</span> <span class="op">&lt;-</span> <span class="st">"/home/yours/an_empty_workdirectory"</span></span></code></pre></div>
<div class="section level2">
<h2 id="preprocessing">Preprocessing<a class="anchor" aria-label="anchor" href="#preprocessing"></a>
</h2>
<p>Count sketching relies on the following ideas: * Genomes can be
represented as a vector of KMER counts. If these KMERs are long enough,
they exist nearly-only once in a genome, and seldom in genomes of
different origins * Angles between KMER count vectors ignore sequencing
depth and thus represent underlying genome prior to
sampling-by-sequencing * The distance between high dimensional vectors
is preserved also after random linear projection into a lower
dimensional space (lower, but still high-dimensional) * The random
projection matrix can be represented implicitly, avoiding the need for
the original full KMER vectors</p>
<p>Note that these linear projections are different from Min-Hashes, and
have further nice properties beyond the description here.</p>
<p>To first compute the project KMER counts, we will use the
BascetGatherCountSketch-command. The length of KMERs, and number of
reduced dimensions, is set at this stage * 31 bp is a common choice of
KMER lengths as it fits well in 32-bit registers of a CPU, while being
rather unique. <em>The KMER length is typically odd</em>, to remove
KMERs symmetric under reverse complement (e.g., AATT) * <em>The reduced
number of dimensions must be a power of two</em>, such as 2**n. Or:
2048, 4096, 8192, 16384 etc. This restriction is due to the choice of
underlying algorithm.</p>
<p>The following command computes a countsketch matrix. The input is
here the raw reads, which is suitable for metatranscriptomics, and works
for metagenomics as well. For higher precision, you can also use the
<em>de novo</em> assembly commands to extract unique KMERs as input for
sketching (not shown here).</p>
<p><a href="slurm.html">(SLURM-compatible step)</a></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Gather count sketches into a single matrix file</span></span>
<span><span class="fu"><a href="../reference/BascetGatherCountSketch.html">BascetGatherCountSketch</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>,</span>
<span>  kmerSize<span class="op">=</span><span class="fl">31</span>,      <span class="co">#Default</span></span>
<span>  sketchSize<span class="op">=</span><span class="fl">4096</span>,  <span class="co">#Need to be power of two</span></span>
<span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="postprocessing-with-signacseurat">Postprocessing with Signac/Seurat<a class="anchor" aria-label="anchor" href="#postprocessing-with-signacseurat"></a>
</h2>
<p>You can now load all of this data into R as a Seurat object:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/BascetLoadCountSketchMatrix.html">BascetLoadCountSketchMatrix</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The data is rather large; we recommend enabling multithreading for
Seurat. <em>Note that in the current version of R/Bioconductor, as of
writing, threading might be disabled in RStudio. You then need to run
the code in a separate R session. If you see warning messages about
threading later on, then this is the only backup we can offer for
now!</em></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org" class="external-link">future</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="st">"multicore"</span>, workers <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p>Note that the data was loaded directly as a <em>Seurat
reduction</em>, rather than as regular counts. This is because the
counts have little meaning of their own, and it makes little sense to
perform PCA, SVD or similar on the data. Rather, we recommend using all
the dimensions, and compute fewer in the first steps if you want to
reduce. The following code produces a UMAP, setting the dimension to all
available dimensions in the reduction:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reduction_name</span> <span class="op">&lt;-</span> <span class="st">"kmersketch"</span></span>
<span><span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span></span>
<span>  <span class="va">adata</span>,</span>
<span>  dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">adata</span><span class="op">@</span><span class="va">reductions</span><span class="op">[[</span><span class="va">reduction_name</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">cell.embeddings</span><span class="op">)</span>,</span>
<span>  reduction <span class="op">=</span> <span class="va">reduction_name</span>,</span>
<span>  metric <span class="op">=</span> <span class="st">"cosine"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The result can be visualized:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span></span></code></pre></div>
<p>We recommend that you integrate this object with the output of
KRAKEN2 to get some clue about what the clusters mean.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Johan Henriksson, Laura Carroll, Hadrien Gourlé, Julian Dicken, HenLab, CompMicroLab, and Collaborators.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
