<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Debarcoding and sharding • Zorn</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Debarcoding and sharding">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Zorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/get_started.html">Get started</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><h6 class="dropdown-header" data-toc-skip>Introductory Vignettes</h6></li>
    <li><a class="dropdown-item" href="../articles/debarcoding.html">Debarcoding</a></li>
    <li><a class="dropdown-item" href="../articles/kraken.html">KRAKEN-based workflow</a></li>
    <li><a class="dropdown-item" href="../articles/alignment.html">Alignment-based workflow</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/fileformats.html">Overview of file formats</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/faq.html">FAQ</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://henlab.org/"><span class="fa fa-home"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/henriksson-lab/zorn"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Debarcoding and sharding</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/henriksson-lab/zorn/blob/main/vignettes/debarcoding.Rmd" class="external-link"><code>vignettes/debarcoding.Rmd</code></a></small>
      <div class="d-none name"><code>debarcoding.Rmd</code></div>
    </div>

    
    
<p>First set up your Zorn/Bascet working directory as before. If you
wish to run these steps on a SLURM cluster, see separate vignette and
adapt accordingly.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/henriksson-lab/zorn" class="external-link">Zorn</a></span><span class="op">)</span></span>
<span><span class="va">bascet_runner</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LocalRunner.html">LocalRunner</a></span><span class="op">(</span>direct <span class="op">=</span> <span class="cn">TRUE</span>, show_script<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">bascetRoot</span> <span class="op">=</span> <span class="st">"/home/yours/an_empty_workdirectory"</span></span></code></pre></div>
<p>The first step in data processing is take the raw FASTQ files and
identify which read belongs to which cell. Zorn can automatically figure
out which files to use for what:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rawmeta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DetectRawFileMeta.html">DetectRawFileMeta</a></span><span class="op">(</span><span class="st">"/home/yours/directory_with_raw_fastq"</span><span class="op">)</span></span></code></pre></div>
<p>TODO explain this data frame</p>
<p>You can now parse the reads the figure out which cell they belong to,
and any UMI information. The only thing you additionally need to specify
is what type of data you have, where “atrandi_wgs” is for the Atrandi
SPC-based MDA protocol for WGS.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/BascetGetRaw.html">BascetGetRaw</a></span><span class="op">(</span></span>
<span>    <span class="va">bascetRoot</span>, </span>
<span>    <span class="va">rawmeta</span>,</span>
<span>    chemistry<span class="op">=</span><span class="st">"atrandi_wgs"</span>,</span>
<span>    runner<span class="op">=</span><span class="va">bascet_runner</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Once barcodes have been processed, you can perform knee plot
analysis, which is the first important quality control.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Decide cells to include</span></span>
<span><span class="va">h</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ReadHistogram.html">ReadHistogram</a></span><span class="op">(</span><span class="va">bascetRoot</span>,<span class="st">"debarcoded"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/PlotHistogram.html">PlotHistogram</a></span><span class="op">(</span><span class="va">h</span><span class="op">)</span></span></code></pre></div>
<p>This is just a histogram of reads per barcode. It however tells you *
how many cells you have * how much free-floating RNA/DNA you have in
your background (empty droplets/SPCs) * how distinct cells are from
background</p>
<p>It makes sense to exclude barcodes if the number of counts are too
low. You can exclude them also at a later stage, but removing the worst
offenders already will speed up computation. Here is one example for
getting the names and number of cells having at least 100 reads:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">includeCells</span> <span class="op">&lt;-</span> <span class="va">h</span><span class="op">$</span><span class="va">cellid</span><span class="op">[</span><span class="va">h</span><span class="op">$</span><span class="va">count</span><span class="op">&gt;</span><span class="fl">100</span><span class="op">]</span>   </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">includeCells</span><span class="op">)</span></span></code></pre></div>
<p>Finally, before you are ready to run all other commands, you need to
make sure that cells are sharded. This is primarily relevant if you have
multiple FASTQ-files for your library (i.e. you sequenced your library
on multiple lanes). Sharding sorts cells such that the reads for one
cell are in one single shard. This is essential for parallel
processing.</p>
<p>At this step you can also choose to create multiple shards. This is
only relevant if you wish to parallelize the workflow using SLURM, as
the number of shards decides how many nodes you can use in parallel. If
you run everything on your own computer then we advise only creating one
shard, as there is no benefit to having multiple and it is more likely
to cause mistakes.</p>
<p>In this step we also filter out barcodes with too low read counts
(optional).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/BascetShardify.html">BascetShardify</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>,</span>
<span>  includeCells <span class="op">=</span> <span class="va">includeCells</span>,</span>
<span>  num_output_shards <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  runner<span class="op">=</span><span class="va">bascet_runner</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>If you make it here then you have passed one of the heaviest and most
critical steps in the workflow!</p>
<p>How you proceed depends on your usecase. If you are doing single-cell
metagenomics of a sample of unknown composition then we recommend the
KRAKEN2 workflow next.</p>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Johan Henriksson, Laura Carroll, Hadrien Gourlé, Julian Dicken, Henriksson Lab, Carroll Lab and Collaborators.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
