<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Debarcoding and sharding • Zorn</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Debarcoding and sharding">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Zorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/get_started.html">Get started</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><h6 class="dropdown-header" data-toc-skip>Introductory Vignettes</h6></li>
    <li><a class="dropdown-item" href="../articles/slurm.html">Using SLURM</a></li>
    <li><a class="dropdown-item" href="../articles/debarcoding.html">Debarcoding</a></li>
    <li><a class="dropdown-item" href="../articles/read_quality_control.html">Read-based quality control</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/kraken.html">KRAKEN-based workflow</a></li>
    <li><a class="dropdown-item" href="../articles/alignment.html">Alignment-based workflow</a></li>
    <li><a class="dropdown-item" href="../articles/kmer.html">Informative KMER-based workflow</a></li>
    <li><a class="dropdown-item" href="../articles/countsketch.html">Count sketch KMER-based workflow</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/clustering.html">Clustering</a></li>
    <li><a class="dropdown-item" href="../articles/assembly.html">De novo assembly</a></li>
    <li><a class="dropdown-item" href="../articles/genome_annotation.html">Genome annotation</a></li>
    <li><a class="dropdown-item" href="../articles/snp_analysis.html">SNP analysis</a></li>
    <li><a class="dropdown-item" href="../articles/map_scripts.html">Map scripts</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/fileformats.html">Overview of file formats</a></li>
    <li><a class="dropdown-item" href="../articles/bascet_files.html">Working with Bascet files</a></li>
    <li><a class="dropdown-item" href="../articles/for_developers.html">For developers</a></li>
    <li><a class="dropdown-item" href="../articles/example_data.html">Example data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../articles/faq.html">FAQ</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://henlab.org/" aria-label="Lab website"><span class="fa fa-home"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/henriksson-lab/zorn" aria-label="GitHub repository"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Debarcoding and sharding</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/henriksson-lab/zorn/blob/main/vignettes/debarcoding.Rmd" class="external-link"><code>vignettes/debarcoding.Rmd</code></a></small>
      <div class="d-none name"><code>debarcoding.Rmd</code></div>
    </div>

    
    
<p>First set up your Zorn/Bascet working directory as before. If you
wish to run these steps on a SLURM cluster, see separate vignette and
adapt accordingly.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/henriksson-lab/zorn" class="external-link">Zorn</a></span><span class="op">)</span></span>
<span><span class="va">bascet_runner.default</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LocalRunner.html">LocalRunner</a></span><span class="op">(</span>direct <span class="op">=</span> <span class="cn">TRUE</span>, showScript<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">bascetRoot</span> <span class="op">&lt;-</span> <span class="st">"/home/yours/an_empty_workdirectory"</span></span></code></pre></div>
<p>The first step in data processing is take the raw FASTQ files and
identify which read belongs to which cell. Zorn can automatically figure
out which files to use for what:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rawmeta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DetectRawFileMeta.html">DetectRawFileMeta</a></span><span class="op">(</span><span class="st">"/home/yours/directory_with_raw_fastq"</span><span class="op">)</span></span></code></pre></div>
<p>Depending on your input data, rawmeta might look like this:</p>
<table class="table">
<caption>Example input file metadata</caption>
<colgroup>
<col width="11%">
<col width="28%">
<col width="28%">
<col width="31%">
</colgroup>
<thead><tr class="header">
<th align="left">prefix</th>
<th align="left">r1</th>
<th align="left">r2</th>
<th align="left">dir</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">UDI_Plate_WT_29</td>
<td align="left">UDI_Plate_WT_29_S43_L007_R1_001.fastq.gz</td>
<td align="left">UDI_Plate_WT_29_S43_L007_R2_001.fastq.gz</td>
<td align="left">/home/m/mahogny/mystore/dataset/250611/fastq</td>
</tr>
<tr class="even">
<td align="left">UDI_Plate_WT_29</td>
<td align="left">UDI_Plate_WT_29_S43_L008_R1_001.fastq.gz</td>
<td align="left">UDI_Plate_WT_29_S43_L008_R2_001.fastq.gz</td>
<td align="left">/home/m/mahogny/mystore/dataset/250611/fastq</td>
</tr>
<tr class="odd">
<td align="left">UDI_Plate_WT_30</td>
<td align="left">UDI_Plate_WT_30_S44_L007_R1_001.fastq.gz</td>
<td align="left">UDI_Plate_WT_30_S44_L007_R2_001.fastq.gz</td>
<td align="left">/home/m/mahogny/mystore/dataset/250611/fastq</td>
</tr>
<tr class="even">
<td align="left">UDI_Plate_WT_30</td>
<td align="left">UDI_Plate_WT_30_S44_L008_R1_001.fastq.gz</td>
<td align="left">UDI_Plate_WT_30_S44_L008_R2_001.fastq.gz</td>
<td align="left">/home/m/mahogny/mystore/dataset/250611/fastq</td>
</tr>
</tbody>
</table>
<p>Prefix is the name of the library, of which there are two here. Note
the paired input FASTQ files R1 and R2. Zorn has here detected that the
libraries were sequenced on two separate lanes; by using the same
prefix, these will be merged during the later sharding stage. Keeping
input files from different lanes separate enables each FASTQ file to be
processed on a separate computer and can thus be advantageous if you
have large amounts of data.</p>
<p>If you have files that are awkwardly named for some reason, simply
produce a data frame with the columns above, and provide as input for
the next stage.</p>
<p>You can now parse the reads the figure out which cell they belong to,
and any UMI information. The only thing you additionally need to specify
is what type of data you have, where “atrandi_wgs” is for the Atrandi
SPC-based MDA protocol for WGS.</p>
<p><a href="slurm.html">(SLURM-compatible step)</a></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/BascetGetRaw.html">BascetGetRaw</a></span><span class="op">(</span></span>
<span>    <span class="va">bascetRoot</span>, </span>
<span>    <span class="va">rawmeta</span>,</span>
<span>    chemistry<span class="op">=</span><span class="st">"atrandi_wgs"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>After debarcoding, you now need to (1) merge files belong to the same
library, and (2) remove reads belonging to empty droplets. Furthermore,
at this step you can also decide to split the cells across multiple
“shards”, enabling parallel processing the data in later steps. This
done in two steps, in which you first set up a “plan”. This is done
using the following command:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">debstat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PrepareSharding.html">PrepareSharding</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>,</span>
<span>  inputName<span class="op">=</span><span class="st">"debarcoded"</span>,</span>
<span>  bascetInstance<span class="op">=</span><span class="va">bascet_instance.default</span>,</span>
<span>  minQuantile<span class="op">=</span><span class="fl">0.99</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Before you proceed, you should perform a kneeplot analysis to see
which cells you keep:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/DebarcodedKneePlot.html">DebarcodedKneePlot</a></span><span class="op">(</span></span>
<span>  <span class="va">debstat</span>,</span>
<span>  filename <span class="op">=</span> <span class="st">"kneeplot.pdf"</span> <span class="co">#optional</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>This is just a histogram of reads per barcode. (downsampling for
speed causes cells of low indices to not be shown, worry not!) The
didagram however tells you * how many cells you have * how much
free-floating RNA/DNA you have in your background (empty droplets/SPCs)
* how distinct cells are from background</p>
<p>An example kneeplot for Parse biosciences RNA-seq data is shown
below. You want to to include slightly more cells than the knee
indicates. For Atrandi microbial WGS data, knee plots cannot be trusted,
and you may want a fair bit more to account for lysis differences (see
our manuscript). Including too many cells will slow down later steps but
you can reduce the number of cells at any point. Adjust the cutoff using
the previously set minQuantile parameter.</p>
<div class="float">
<img src="debarcoding_kneeplot.png" alt="Kneeplot"><div class="figcaption">Kneeplot</div>
</div>
<p>Once you have a plan you are happy with, you can proceed with
shardification. At this step you can also choose to create multiple
shards. This is only relevant if you wish to parallelize the workflow
using SLURM, as the number of shards decides how many nodes you can use
in parallel. If you run everything on your own computer then we advise
only creating one shard, as there is no benefit to having multiple and
it is more likely to cause mistakes.</p>
<p><a href="slurm.html">(SLURM-compatible step)</a></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/BascetShardify.html">BascetShardify</a></span><span class="op">(</span></span>
<span>  <span class="va">debstat</span>,</span>
<span>  numOutputShards <span class="op">=</span> <span class="fl">1</span>, <span class="co">#shards per library; increase to spread the workload over more computers</span></span>
<span>  runner<span class="op">=</span><span class="fu"><a href="../reference/SlurmRunner.html">SlurmRunner</a></span><span class="op">(</span><span class="va">bascet_runner.default</span>, ncpu<span class="op">=</span><span class="st">"4"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="additional-trimming">Additional trimming<a class="anchor" aria-label="anchor" href="#additional-trimming"></a>
</h2>
<p>Bascet only does rudimentary trimming. We recommend trimming the
reads further. To do this, first convert your data to FASTQ format:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Get reads in fastq format</span></span>
<span><span class="fu"><a href="../reference/BascetMapTransform.html">BascetMapTransform</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>,</span>
<span>  inputName<span class="op">=</span><span class="st">"filtered"</span>,   <span class="co">#default; can omit</span></span>
<span>  outputName<span class="op">=</span><span class="st">"asfq"</span>,</span>
<span>  out_format<span class="op">=</span><span class="st">"R1.fq.gz"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Now use FASTP to perform the trimming:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Get reads in fastq format for BWA</span></span>
<span><span class="fu"><a href="../reference/BascetRunFASTP.html">BascetRunFASTP</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>,</span>
<span>  numLocalThreads<span class="op">=</span><span class="fl">10</span>,</span>
<span>  inputName<span class="op">=</span><span class="st">"asfq"</span>,</span>
<span>  outputName<span class="op">=</span><span class="st">"fastp"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>If your plan is to perform alignment, then the resulting fastp FASTQ
files can be used as input (don’t forget to specify inputName=“fastp”!).
If you want to instead get a TIRP file, you can convert it back:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/BascetMapTransform.html">BascetMapTransform</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>,</span>
<span>  <span class="st">"fastp"</span>,</span>
<span>  <span class="st">"new_filtered"</span>,</span>
<span>  out_format<span class="op">=</span><span class="st">"tirp.gz"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>You now have a TIRP file again, equivalent to the first shardified
output. Don’t forget to specify inputName=“new_filtered” to later
commands, as the default is to use “filtered”</p>
</div>
<div class="section level2">
<h2 id="onward">Onward<a class="anchor" aria-label="anchor" href="#onward"></a>
</h2>
<p>If you make it here then you have passed one of the heaviest and most
critical steps in the workflow!</p>
<p>How you proceed depends on your use case. If you are doing
single-cell metagenomics of a sample of unknown composition then we
recommend the KRAKEN2 workflow next.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Johan Henriksson, Laura Carroll, Hadrien Gourlé, Julian Dicken, HenLab, CompMicroLab, and Collaborators.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
