# Articles

### All vignettes

- [Alignment-based
  workflow](https://henriksson-lab.github.io/zorn/articles/alignment.md):
- [De novo
  assembly](https://henriksson-lab.github.io/zorn/articles/assembly.md):
- [Working with Bascet
  files](https://henriksson-lab.github.io/zorn/articles/bascet_files.md):
- [Clustering](https://henriksson-lab.github.io/zorn/articles/clustering.md):
- [Count sketch KMER-based
  workflow](https://henriksson-lab.github.io/zorn/articles/countsketch.md):
- [Debarcoding and
  sharding](https://henriksson-lab.github.io/zorn/articles/debarcoding.md):
- [Example
  data](https://henriksson-lab.github.io/zorn/articles/example_data.md):
- [Frequently Asked
  Questions](https://henriksson-lab.github.io/zorn/articles/faq.md):
- [Overview of File
  formats](https://henriksson-lab.github.io/zorn/articles/fileformats.md):
- [Installation](https://henriksson-lab.github.io/zorn/articles/for_developers.md):
- [Genome annotation
  tools](https://henriksson-lab.github.io/zorn/articles/genome_annotation.md):
- [Getting Started with
  Zorn](https://henriksson-lab.github.io/zorn/articles/get_started.md):
- [Installation](https://henriksson-lab.github.io/zorn/articles/install.md):
- [Isolate
  analysis](https://henriksson-lab.github.io/zorn/articles/isolates.md):
- [Informative KMER-based
  workflow](https://henriksson-lab.github.io/zorn/articles/kmer.md):
- [KRAKEN2](https://henriksson-lab.github.io/zorn/articles/kraken.md) :
- [Map
  scripts](https://henriksson-lab.github.io/zorn/articles/map_scripts.md):
- [Read-based quality
  control](https://henriksson-lab.github.io/zorn/articles/read_quality_control.md):
- [Using
  SLURM](https://henriksson-lab.github.io/zorn/articles/slurm.md):
- [SNP
  analysis](https://henriksson-lab.github.io/zorn/articles/snp_analysis.md):
- [TCR/BCR
  analysis](https://henriksson-lab.github.io/zorn/articles/tcr_analysis.md):
