<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Informative KMER-based workflow • Zorn</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Informative KMER-based workflow">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Zorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/get_started.html">Get started</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><h6 class="dropdown-header" data-toc-skip>Introductory Vignettes</h6></li>
    <li><a class="dropdown-item" href="../articles/slurm.html">Using SLURM</a></li>
    <li><a class="dropdown-item" href="../articles/debarcoding.html">Debarcoding</a></li>
    <li><a class="dropdown-item" href="../articles/read_quality_control.html">Read-based quality control</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/kraken.html">KRAKEN-based workflow</a></li>
    <li><a class="dropdown-item" href="../articles/alignment.html">Alignment-based workflow</a></li>
    <li><a class="dropdown-item" href="../articles/kmer.html">Informative KMER-based workflow</a></li>
    <li><a class="dropdown-item" href="../articles/countsketch.html">Count sketch KMER-based workflow</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/clustering.html">Clustering</a></li>
    <li><a class="dropdown-item" href="../articles/assembly.html">De novo assembly</a></li>
    <li><a class="dropdown-item" href="../articles/genome_annotation.html">Genome annotation</a></li>
    <li><a class="dropdown-item" href="../articles/snp_analysis.html">SNP analysis</a></li>
    <li><a class="dropdown-item" href="../articles/map_scripts.html">Map scripts</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/fileformats.html">Overview of file formats</a></li>
    <li><a class="dropdown-item" href="../articles/bascet_files.html">Working with Bascet files</a></li>
    <li><a class="dropdown-item" href="../articles/for_developers.html">For developers</a></li>
    <li><a class="dropdown-item" href="../articles/example_data.html">Example data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../articles/faq.html">FAQ</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://henlab.org/" aria-label="Lab website"><span class="fa fa-home"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/henriksson-lab/zorn" aria-label="GitHub repository"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Informative KMER-based workflow</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/henriksson-lab/zorn/blob/main/vignettes/kmer.Rmd" class="external-link"><code>vignettes/kmer.Rmd</code></a></small>
      <div class="d-none name"><code>kmer.Rmd</code></div>
    </div>

    
    
<p>First set up your Zorn/Bascet workdirectory as before. If you wish to
run these steps on a SLURM cluster, see separate vignette and adapt
accordingly.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/henriksson-lab/zorn" class="external-link">Zorn</a></span><span class="op">)</span></span>
<span><span class="va">bascet_runner.default</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LocalRunner.html">LocalRunner</a></span><span class="op">(</span>direct <span class="op">=</span> <span class="cn">TRUE</span>, showScript<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">bascetRoot</span> <span class="op">&lt;-</span> <span class="st">"/home/yours/an_empty_workdirectory"</span></span></code></pre></div>
<div class="section level2">
<h2 id="preprocessing">Preprocessing<a class="anchor" aria-label="anchor" href="#preprocessing"></a>
</h2>
<p>To extract informative KMERs, you first need to obtain a list of
them. Bascet/Zorn offers one way, in which we first extract minhashes
from each cell. These are finger prints that we later can look for in
other cells.</p>
<p><a href="slurm.html">(SLURM-compatible step)</a></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Compute minhashes for each cell</span></span>
<span><span class="fu"><a href="../reference/BascetComputeMinhash.html">BascetComputeMinhash</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Gather minhashes into a single histogram</span></span>
<span><span class="fu"><a href="../reference/BascetMakeMinhashHistogram.html">BascetMakeMinhashHistogram</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>If two cells of the same species are shallowly sequenced, they may
not share minhashes, simply because the reads don’t span all of their
genomes. We can however look for KMERs that are more frequently shared
by counting minhashes across all cells, and generating a histogram.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kmerHist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/BascetReadMinhashHistogram.html">BascetReadMinhashHistogram</a></span><span class="op">(</span><span class="va">bascetRoot</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kmerHist</span><span class="op">$</span><span class="va">rank</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">kmerHist</span><span class="op">)</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">kmerHist</span><span class="op">[</span><span class="va">kmerHist</span><span class="op">$</span><span class="va">cnt</span><span class="op">&gt;</span><span class="fl">2</span>,<span class="op">]</span>, <span class="fu">aes</span><span class="op">(</span><span class="va">rank</span>, <span class="va">cnt</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_x_log10</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_y_log10</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>It is an open problem which KMERs to pick. KMERs of low abundance
will not help clustering much as the UMAP algorithm will not find many
other cells they are correlated with. Likewise, KMERs present in each
cell are not <a href="https://en.wikipedia.org/wiki/Binary_entropy_function" class="external-link">informative
either</a>. In practice, we don’t see this happening; just picking the
most common top 10,000 KMERs or so seems to give useful results. You can
also pick KMERs based on a cutoff:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Pick KMERs</span></span>
<span><span class="va">useKMERs</span> <span class="op">&lt;-</span> <span class="va">kmerHist</span><span class="op">$</span><span class="va">kmer</span><span class="op">[</span><span class="va">kmerHist</span><span class="op">$</span><span class="va">cnt</span><span class="op">&gt;</span><span class="fl">5</span><span class="op">]</span>  <span class="co">#Pick KMERs present in more than 5 cells</span></span></code></pre></div>
<p>Using the list of KMERs, you can now re-scan all genomes for their
presence.</p>
<p><a href="slurm.html">(SLURM-compatible step)</a></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Build count table by looking up selected KMERs in per-cell KMER databases</span></span>
<span><span class="fu"><a href="../reference/BascetQueryFq.html">BascetQueryFq</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>,</span>
<span>  useKMERs<span class="op">=</span><span class="va">useKMERs</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The output format is one binary HDF5 file for each shard, roughly in
the <a href="https://anndata.readthedocs.io/en/stable/fileformat-prose.html" class="external-link">Anndata
file format</a>. To load these, use the following command that loads the
files and concatenates them into a single matrix. It can then be loaded
into Seurat:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="co">### Read count matrix</span></span>
<span><span class="va">cnt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ReadBascetCountMatrix.html">ReadBascetCountMatrix</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>,</span>
<span>  <span class="st">"kmer_counts"</span></span>
<span>  <span class="co">#TODO note that the is a maximum count by default</span></span>
<span><span class="op">)</span> </span></code></pre></div>
<p>Most likely you will have to filter out the low abundant cells. Note
that with this workflow, the cutoff is based on the number of KMERs
rather than the number of reads.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co">## Load subset of real cells</span></span>
<span>  <span class="va">keep_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">cnt</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">20000</span> <span class="co">#this cutoff is based on the number of KMERs, not read count!</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">keep_cells</span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span></span>
<span>    counts <span class="op">=</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/CreateAssayObject.html" class="external-link">CreateAssayObject</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">cnt</span><span class="op">[</span><span class="va">keep_cells</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span>,  <span class="co">#note that t() is needed</span></span>
<span>    assay <span class="op">=</span> <span class="st">"infokmer"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>You can now proceed to generate a dimensional reduction in the form
of a UMAP. See the tutorials for <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a> and <a href="https://stuartlab.org/signac/" class="external-link">Signac</a> for details. We
summarize the required command below, where you can apply either an
RNA-seq or ATAC-seq dimensional reduction depending on what you think is
appropriate.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co">#ATAC-seq style analysis</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stuart-lab/signac" class="external-link">Signac</a></span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/RunTFIDF.html" class="external-link">RunTFIDF</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/FindTopFeatures.html" class="external-link">FindTopFeatures</a></span><span class="op">(</span><span class="va">adata</span>, min.cutoff <span class="op">=</span> <span class="st">'q0'</span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/RunSVD.html" class="external-link">RunSVD</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://stuartlab.org/signac/reference/DepthCor.html" class="external-link">DepthCor</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">adata</span>, reduction <span class="op">=</span> <span class="st">'lsi'</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, reduction.name <span class="op">=</span> <span class="st">"infokmers_umap"</span><span class="op">)</span>  </span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="co">#RNA-seq style analysis</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">adata</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span>, nfeatures <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">adata</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">adata</span>, features <span class="op">=</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">adata</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">adata</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, reduction.name <span class="op">=</span> <span class="st">"infokmers_umap"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>You can now plot a UMAP with your cells. Note that the cells do not
have any labels. One way to obtain labels is by also running the
KRAKEN-based workflow; or you can use other annotation tools through the
mapcell-system.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">adata</span>, reduction <span class="op">=</span> <span class="st">"infokmers_umap"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">xlab</span><span class="op">(</span><span class="st">"KMER1"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ylab</span><span class="op">(</span><span class="st">"KMER2"</span><span class="op">)</span></span></code></pre></div>
</div>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Johan Henriksson, Laura Carroll, Hadrien Gourlé, Julian Dicken, Henriksson Lab, Carroll Lab and Collaborators.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
