<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>KRAKEN2  • Zorn</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="KRAKEN2 ">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Zorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/get_started.html">Get started</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><h6 class="dropdown-header" data-toc-skip>Introductory Vignettes</h6></li>
    <li><a class="dropdown-item" href="../articles/slurm.html">Using SLURM</a></li>
    <li><a class="dropdown-item" href="../articles/debarcoding.html">Debarcoding</a></li>
    <li><a class="dropdown-item" href="../articles/read_quality_control.html">Read-based quality control</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/kraken.html">KRAKEN-based workflow</a></li>
    <li><a class="dropdown-item" href="../articles/alignment.html">Alignment-based workflow</a></li>
    <li><a class="dropdown-item" href="../articles/kmer.html">Informative KMER-based workflow</a></li>
    <li><a class="dropdown-item" href="../articles/countsketch.html">Count sketch KMER-based workflow</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/clustering.html">Clustering</a></li>
    <li><a class="dropdown-item" href="../articles/assembly.html">De novo assembly</a></li>
    <li><a class="dropdown-item" href="../articles/genome_annotation.html">Genome annotation</a></li>
    <li><a class="dropdown-item" href="../articles/snp_analysis.html">SNP analysis</a></li>
    <li><a class="dropdown-item" href="../articles/map_scripts.html">Map scripts</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/fileformats.html">Overview of file formats</a></li>
    <li><a class="dropdown-item" href="../articles/bascet_files.html">Working with Bascet files</a></li>
    <li><a class="dropdown-item" href="../articles/for_developers.html">For developers</a></li>
    <li><a class="dropdown-item" href="../articles/example_data.html">Example data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../articles/faq.html">FAQ</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://henlab.org/" aria-label="Lab website"><span class="fa fa-home"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/henriksson-lab/zorn" aria-label="GitHub repository"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>KRAKEN2 </h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/henriksson-lab/zorn/blob/main/vignettes/kraken.Rmd" class="external-link"><code>vignettes/kraken.Rmd</code></a></small>
      <div class="d-none name"><code>kraken.Rmd</code></div>
    </div>

    
    
<p>First set up your Zorn/Bascet working directory as before. If you
wish to run these steps on a SLURM cluster, see separate vignette and
adapt accordingly.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/henriksson-lab/zorn" class="external-link">Zorn</a></span><span class="op">)</span></span>
<span><span class="va">bascet_runner.default</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LocalRunner.html">LocalRunner</a></span><span class="op">(</span>direct <span class="op">=</span> <span class="cn">TRUE</span>, showScript<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">bascetRoot</span> <span class="op">&lt;-</span> <span class="st">"/home/yours/an_empty_workdirectory"</span></span></code></pre></div>
<div class="section level2">
<h2 id="preprocessing">Preprocessing<a class="anchor" aria-label="anchor" href="#preprocessing"></a>
</h2>
<p>KRAKEN2 wants the data in FASTQ format. Here we assume that you use a
single-cell WGS chemistry that produces paired reads, and direct Zorn to
produce R1/R2 FASTQ files (only R1 specified). We name the output
Bascets “asfq”, taking as input the typical sharded reads:</p>
<p><a href="slurm.html">(SLURM-compatible step)</a></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Get reads in fastq format</span></span>
<span><span class="fu"><a href="../reference/BascetMapTransform.html">BascetMapTransform</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>,</span>
<span>  <span class="st">"filtered"</span>,   <span class="co">#default; can omit</span></span>
<span>  <span class="st">"asfq"</span>,  <span class="co">###name parameter</span></span>
<span>  out_format<span class="op">=</span><span class="st">"R1.fq.gz"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>To run KRAKEN2, you need a database. You can get them here: <a href="https://benlangmead.github.io/aws-indexes/k2" class="external-link uri">https://benlangmead.github.io/aws-indexes/k2</a></p>
<p>If you want a small one then consider standard-8. Unzip it in a
directory. You can then run KRAKEN2 like this:</p>
<p><a href="slurm.html">(SLURM-compatible step)</a></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Run Kraken on each cell</span></span>
<span><span class="fu"><a href="../reference/BascetRunKraken.html">BascetRunKraken</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>,</span>
<span>  useKrakenDB<span class="op">=</span><span class="st">"/your_disk/kraken/standard-8"</span>,</span>
<span>  numLocalThreads<span class="op">=</span><span class="fl">20</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Produce a count matrix of taxonomy features</span></span>
<span><span class="fu"><a href="../reference/BascetMakeKrakenCountMatrix.html">BascetMakeKrakenCountMatrix</a></span><span class="op">(</span></span>
<span>  <span class="va">bascetRoot</span>,</span>
<span>  numLocalThreads<span class="op">=</span><span class="fl">20</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Note that there are two steps here. First KRAKEN2 classifies each
read to a taxonomy. In the second step, we count the taxonomic reads for
each cell. This ends up being a rather small matrix that you can process
using Seurat.</p>
</div>
<div class="section level2">
<h2 id="postprocessing-with-signacseurat">Postprocessing with Signac/Seurat<a class="anchor" aria-label="anchor" href="#postprocessing-with-signacseurat"></a>
</h2>
<p>The output format is one binary HDF5 file for each shard, roughly in
the <a href="https://anndata.readthedocs.io/en/stable/fileformat-prose.html" class="external-link">Anndata
file format</a>. To load these, use the following command that loads the
files and concatenates them into a single matrix. It can then be loaded
into Seurat:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ReadBascetCountMatrix.html">ReadBascetCountMatrix</a></span><span class="op">(</span><span class="va">bascetRoot</span>,<span class="st">"kraken"</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">taxid_ob</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/CreateAssayObject.html" class="external-link">CreateAssayObject</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span><span class="op">)</span> <span class="co">#t() needed to conform from anndata to seurat</span></span>
<span><span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">taxid_ob</span>, project <span class="op">=</span> <span class="st">"proj"</span>, min.cells <span class="op">=</span> <span class="fl">0</span>, min.features <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> </span></code></pre></div>
<p>You then most likely want to know which cell is which species. Zorn
can annotate species by simply picking the dominant species, phylum and
group for each cell:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Look up taxonomy consensus data given taxonomyID counts for each cell</span></span>
<span><span class="va">kraken_taxid</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/KrakenFindConsensusTaxonomy.html">KrakenFindConsensusTaxonomy</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Add KRAKEN consensus taxonomy to metadata</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">kraken_taxid</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">kraken_taxid</span><span class="op">$</span><span class="va">cell_id</span></span>
<span><span class="va">kraken_taxid</span> <span class="op">&lt;-</span> <span class="va">kraken_taxid</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"taxid"</span>,<span class="st">"phylum"</span>,<span class="st">"class"</span>,<span class="st">"order"</span>,<span class="st">"family"</span>,<span class="st">"genus"</span>,<span class="st">"species"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">adata</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span></span>
<span>  <span class="va">adata</span><span class="op">@</span><span class="va">meta.data</span>,</span>
<span>  <span class="va">kraken_taxid</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"taxid"</span>,<span class="st">"phylum"</span>,<span class="st">"class"</span>,<span class="st">"order"</span>,<span class="st">"family"</span>,<span class="st">"genus"</span>,<span class="st">"species"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>You will need to filter low-abundance cells. To do this, first
investigate a kneeplot of each species. As different species are
differently hard to lyse, you will likely have different kneeplot
patterns.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/KrakenKneePlot.html">KrakenKneePlot</a></span><span class="op">(</span><span class="va">adata</span>, groupby <span class="op">=</span> <span class="st">"phylum"</span>, showNumSpecies<span class="op">=</span><span class="va">showNumSpecies</span><span class="op">)</span></span></code></pre></div>
<p>You can then proceed to filter out cells with few reads. Note that
the cutoff will induce biases if the cells have different DNA content;
the best way to handle this is an open research question (e.g. you could
also consider different cutoffs for different species). This is however
the most basic filtering you can perform:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keep_cells</span> <span class="op">&lt;-</span> <span class="va">adata</span><span class="op">$</span><span class="va">nCount_RNA</span> <span class="op">&gt;</span> <span class="fl">10000</span> <span class="co">#10k reads</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">keep_cells</span><span class="op">)</span></span>
<span><span class="va">adata</span> <span class="op">&lt;-</span> <span class="va">adata</span><span class="op">[</span>,<span class="va">keep_cells</span><span class="op">]</span></span></code></pre></div>
<p>You can now proceed to generate a dimensional reduction in the form
of a UMAP. See the tutorials for <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a> and <a href="https://stuartlab.org/signac/" class="external-link">Signac</a> for details. We
summarize the required command below, where you can apply either an
RNA-seq or ATAC-seq dimensional reduction depending on what you think is
appropriate.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co">#ATAC-seq style analysis</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stuart-lab/signac" class="external-link">Signac</a></span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/RunTFIDF.html" class="external-link">RunTFIDF</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/FindTopFeatures.html" class="external-link">FindTopFeatures</a></span><span class="op">(</span><span class="va">adata</span>, min.cutoff <span class="op">=</span> <span class="st">'q0'</span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/RunSVD.html" class="external-link">RunSVD</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://stuartlab.org/signac/reference/DepthCor.html" class="external-link">DepthCor</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">adata</span>, reduction <span class="op">=</span> <span class="st">'lsi'</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, reduction.name <span class="op">=</span> <span class="st">"kraken_umap"</span><span class="op">)</span>  </span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="co">#RNA-seq style analysis</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">adata</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span>, nfeatures <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">adata</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">adata</span>, features <span class="op">=</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">adata</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">adata</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, reduction.name <span class="op">=</span> <span class="st">"kraken_umap"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>You can now plot a UMAP with your cells! Here colored by genus, but
you can also try other levels of annotation:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">adata</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, group.by <span class="op">=</span> <span class="st">"genus"</span>, reduction <span class="op">=</span> <span class="st">"kraken_umap"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">xlab</span><span class="op">(</span><span class="st">"KRAKEN1"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ylab</span><span class="op">(</span><span class="st">"KRAKEN2"</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Johan Henriksson, Laura Carroll, Hadrien Gourlé, Julian Dicken, Henriksson Lab, Carroll Lab and Collaborators.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
